// send/lib/mime.js
/*!
 * send
 * Copyright(c) 2012 TJ Holowaychuk
 * Copyright(c) 2014-2022 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var mime = require('mime-types')
var deprecate = require('depd')('send')

/**
 * Module variables.
 * @private
 */

var extensionToMimeType = {
  // Text
  'css': 'text/css',
  'html': 'text/html',
  'js': 'text/javascript',
  'txt': 'text/plain',
  'xml': 'text/xml',
  
  // Images
  'gif': 'image/gif',
  'ico': 'image/x-icon',
  'jpeg': 'image/jpeg',
  'jpg': 'image/jpeg',
  'png': 'image/png',
  'svg': 'image/svg+xml',
  'webp': 'image/webp',
  
  // Audio
  'mp3': 'audio/mpeg',
  'ogg': 'audio/ogg',
  'wav': 'audio/wav',
  
  // Video
  'mp4': 'video/mp4',
  'mpeg': 'video/mpeg',
  'webm': 'video/webm',
  
  // Application
  'json': 'application/json',
  'pdf': 'application/pdf',
  'zip': 'application/zip',
  'gz': 'application/gzip',
  
  // Fonts
  'eot': 'application/vnd.ms-fontobject',
  'ttf': 'font/ttf',
  'woff': 'font/woff',
  'woff2': 'font/woff2',
  
  // Microsoft Office
  'doc': 'application/msword',
  'docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  'xls': 'application/vnd.ms-excel',
  'xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  'ppt': 'application/vnd.ms-powerpoint',
  'pptx': 'application/vnd.openxmlformats-officedocument.presentationml.presentation'
}

/**
 * Default type.
 * @private
 */

var DEFAULT_TYPE = 'application/octet-stream'

/**
 * Get the content type for a given extension.
 *
 * @param {String} extension
 * @return {String}
 * @private
 */

function getType(extension) {
  if (!extension) {
    return DEFAULT_TYPE
  }
  
  // Remove leading dot if present
  if (extension.charAt(0) === '.') {
    extension = extension.substr(1)
  }
  
  // Convert to lowercase for case-insensitive matching
  extension = extension.toLowerCase()
  
  // Look up in our extension map
  var type = extensionToMimeType[extension]
  
  // If not found in our map, try mime-types module
  if (!type && mime) {
    type = mime.lookup(extension)
  }
  
  return type || DEFAULT_TYPE
}

/**
 * Get the extension for a given content type.
 *
 * @param {String} type
 * @return {String}
 * @private
 */

function getExtension(type) {
  if (!type || typeof type !== 'string') {
    return undefined
  }

  // Try mime-types module if available
  if (mime) {
    return mime.extension(type)
  }
  
  // Fallback to manual lookup
  type = type.toLowerCase()
  
  for (var ext in extensionToMimeType) {
    if (extensionToMimeType[ext] === type) {
      return ext
    }
  }
  
  return undefined
}

/**
 * Module exports.
 * @public
 */

exports.charsets = {
  lookup: function(mimeType) {
    // UTF-8 default for text-based content
    if (/^text\/|^application\/json|^application\/xml/.test(mimeType)) {
      return 'UTF-8'
    }
    
    return false
  }
}

/**
 * @deprecated Use mime-types npm module instead
 */

exports.lookup = deprecate(getType,
  'mime.lookup() will be removed in the next major version, use require("mime-types").lookup() instead')

/**
 * @deprecated Use mime-types npm module instead
 */

exports.extension = deprecate(getExtension,
  'mime.extension() will be removed in the next major version, use require("mime-types").extension() instead')

/**
 * Export the content-type for a given extension.
 *
 * @param {String} extension
 * @return {String}
 * @api public
 */

exports.contentType = function contentType(extension) {
  var type = getType(extension)
  
  if (!type) {
    return false
  }
  
  if (!exports.charsets.lookup(type)) {
    return type
  }
  
  return type + '; charset=utf-8'
}

/**
 * Export the MIME type map.
 * @api public
 */

exports.types = Object.create(null)

// Populate types map with our basic mapping
Object.keys(extensionToMimeType).forEach(function(ext) {
  exports.types[ext] = extensionToMimeType[ext]
})

/**
 * Export the default type.
 * @api public
 */

exports.default_type = DEFAULT_TYPE
