// send/lib/utils.js
/**
 * Module dependencies.
 */

var crypto = require('crypto');
var fs = require('fs');
var path = require('path');
var Stream = require('stream');

/**
 * Return strong ETag for `body`.
 *
 * @param {String|Buffer} body
 * @param {String} [encoding]
 * @return {String}
 * @api private
 */

exports.etag = function etag(body, encoding) {
  if (body == null) return undefined;
  var buf = !Buffer.isBuffer(body)
    ? Buffer.from(body, encoding)
    : body;

  return '"' + crypto
    .createHash('sha1')
    .update(buf)
    .digest('hex') + '"';
};

/**
 * Return weak ETag for `stats`.
 *
 * @param {Object} stats
 * @return {String}
 * @api private
 */

exports.wetag = function wetag(stats) {
  if (stats == null) return undefined;
  var mtime = stats.mtime.getTime().toString(16);
  var size = stats.size.toString(16);

  return 'W/"' + size + '-' + mtime + '"';
};

/**
 * Check if the request is cacheable, aka
 * responded with 2xx or 304 (see RFC 2616 section 14.2{5,6}).
 *
 * @param {Number} status
 * @return {Boolean}
 * @api private
 */

exports.isCacheable = function isCacheable(status) {
  return (status >= 200 && status < 300) || status === 304;
};

/**
 * Check if the request is fresh, aka
 * Last-Modified and/or the ETag still match.
 *
 * @param {Object} req
 * @param {Object} res
 * @return {Boolean}
 * @api private
 */

exports.isFresh = function isFresh(req, res) {
  // fields
  var modifiedSince = req.headers['if-modified-since'];
  var noneMatch = req.headers['if-none-match'];

  // unconditional request
  if (!modifiedSince && !noneMatch) return false;

  // Always return stale when Cache-Control: no-cache
  // to support end-to-end reload requests
  // https://tools.ietf.org/html/rfc2616#section-14.9.4
  var cacheControl = req.headers['cache-control'];
  if (cacheControl && cacheControl.indexOf('no-cache') !== -1) {
    return false;
  }

  // if-none-match
  if (noneMatch && noneMatch !== '*') {
    var etag = res.get('ETag');

    if (!etag) {
      return false;
    }

    var etagStale = true;
    var matches = noneMatch.split(/ *, */);
    for (var i = 0; i < matches.length; i++) {
      var match = matches[i];
      if (match === etag || match === 'W/' + etag || 'W/' + match === etag) {
        etagStale = false;
        break;
      }
    }

    if (etagStale) {
      return false;
    }
  }

  // if-modified-since
  if (modifiedSince) {
    var lastModified = res.get('Last-Modified');
    var modifiedStale = !lastModified || !(new Date(lastModified) <= new Date(modifiedSince));

    if (modifiedStale) {
      return false;
    }
  }

  return true;
};

/**
 * Check if the range is fresh.
 *
 * @param {Object} req
 * @param {Object} res
 * @return {Boolean}
 * @api private
 */

exports.isRangeFresh = function isRangeFresh(req, res) {
  var ifRange = req.headers['if-range'];

  if (!ifRange) return true;

  // if-range as etag
  if (ifRange.indexOf('"') !== -1) {
    var etag = res.get('ETag');
    return Boolean(etag && ifRange.indexOf(etag) !== -1);
  }

  // if-range as modified date
  var lastModified = res.get('Last-Modified');
  return parseHttpDate(lastModified) <= parseHttpDate(ifRange);
};

/**
 * Parse an HTTP Date into a number.
 *
 * @param {String} date
 * @return {Number}
 * @api private
 */

function parseHttpDate(date) {
  var timestamp = date && Date.parse(date);

  // istanbul ignore next: guard against date.js Date.parse patching
  return typeof timestamp === 'number'
    ? timestamp
    : NaN;
}

/**
 * Strip content-* header fields.
 *
 * @param {Object} res
 * @api private
 */

exports.removeContentHeaders = function removeContentHeaders(res) {
  var headers = Object.keys(res._headers || {});

  for (var i = 0; i < headers.length; i++) {
    var header = headers[i];
    if (header.substr(0, 8) === 'content-' && header !== 'content-location') {
      res.removeHeader(header);
    }
  }
};

/**
 * Check if `path` looks absolute.
 *
 * @param {String} path
 * @return {Boolean}
 * @api private
 */

exports.isAbsolute = function isAbsolute(path) {
  if (path[0] === '/') return true;
  if (path[1] === ':' && (path[2] === '\\' || path[2] === '/')) return true; // Windows device path
  if (path.substring(0, 2) === '\\\\') return true; // Windows UNC path
  return false;
};

/**
 * Normalize the index option into an array.
 *
 * @param {Boolean|String|Array} val
 * @return {Array}
 * @api private
 */

exports.normalizeList = function normalizeList(val) {
  var list = []

  if (val === false) {
    return list
  }

  if (val === true) {
    list.push('index.html')
    return list
  }

  if (typeof val === 'string') {
    list.push(val)
    return list
  }

  return val || list
};

/**
 * Determine if path parts contain a dotfile.
 *
 * @param {Array} parts
 * @return {Boolean}
 * @api private
 */

exports.containsDotFile = function containsDotFile(parts) {
  for (var i = 0; i < parts.length; i++) {
    var part = parts[i];
    if (part.length > 1 && part[0] === '.') {
      return true;
    }
  }

  return false;
};

/**
 * Create a Content-Range header.
 *
 * @param {String} type
 * @param {Number} size
 * @param {Array} [range]
 * @return {String}
 * @api private
 */

exports.contentRange = function contentRange(type, size, range) {
  return type + ' ' + (range ? range.start + '-' + range.end : '*') + '/' + size;
};

/**
 * Create a Range header.
 *
 * @param {Number} start
 * @param {Number} end
 * @return {String}
 * @api private
 */

exports.rangeHeader = function rangeHeader(start, end) {
  return 'bytes=' + start + '-' + end;
};

/**
 * Parse a HTTP token list.
 *
 * @param {String} str
 * @return {Array}
 * @api private
 */

exports.parseTokenList = function parseTokenList(str) {
  var end = 0;
  var list = [];
  var start = 0;

  // gather tokens
  for (var i = 0, len = str.length; i < len; i++) {
    switch (str.charCodeAt(i)) {
      case 0x20: /*   */
        if (start === end) {
          start = end = i + 1;
        }
        break;
      case 0x2c: /* , */
        list.push(str.substring(start, end));
        start = end = i + 1;
        break;
      default:
        end = i + 1;
        break;
    }
  }

  // final token
  list.push(str.substring(start, end));

  return list;
};

/**
 * Parse a byte range header.
 *
 * @param {String} str
 * @param {Number} size
 * @return {Array}
 * @api private
 */

exports.parseRange = function parseRange(str, size) {
  if (str === undefined || str === null) {
    return;
  }

  var index = str.indexOf('=');

  if (index === -1) {
    return;
  }

  // split the range string
  var arr = str.slice(index + 1).split(',');
  var ranges = [];

  // add ranges type
  ranges.type = str.slice(0, index);

  // parse all ranges
  for (var i = 0; i < arr.length; i++) {
    var range = arr[i].split('-');
    var start = parseInt(range[0], 10);
    var end = parseInt(range[1], 10);

    // -nnn
    if (isNaN(start)) {
      start = size - end;
      end = size - 1;
    // nnn-
    } else if (isNaN(end)) {
      end = size - 1;
    }

    // limit last-byte-pos to current length
    if (end > size - 1) {
      end = size - 1;
    }

    // invalid or unsatisfiable
    if (isNaN(start) || isNaN(end) || start > end || start < 0) {
      continue;
    }

    // add range
    ranges.push({
      start: start,
      end: end
    });
  }

  if (ranges.length < 1) {
    // unsatisfiable
    return;
  }

  return ranges;
};

/**
 * Resolve a virtual path to a fully qualified path.
 *
 * @param {String} rootPath
 * @param {String} virtualPath
 * @return {String}
 * @api private
 */

exports.resolvePath = function resolvePath(rootPath, virtualPath) {
  // resolve relative path
  var resolvedPath = path.resolve(rootPath, virtualPath);

  // ensure path starts with root
  if (resolvedPath.indexOf(rootPath) !== 0) {
    throw new Error('Forbidden');
  }

  return resolvedPath;
};

/**
 * Determine if the response headers contain the content-encoding header
 * and it contains the gzip value.
 *
 * @param {Object} res
 * @return {Boolean}
 * @api private
 */

exports.isGzipped = function isGzipped(res) {
  var contentEncoding = res.getHeader('content-encoding');
  return contentEncoding && contentEncoding.indexOf('gzip') !== -1;
};

/**
 * Determine if the request contains the range header and
 * the range is satisfiable.
 *
 * @param {Object} req
 * @param {Object} res
 * @return {Boolean}
 * @api private
 */

exports.isRangeSatisfiable = function isRangeSatisfiable(req, res) {
  var range = req.headers['range'];
  if (!range) return false;

  // parse
  var ranges = exports.parseRange(range, res.getHeader('content-length'));

  // unsatisfiable
  if (!ranges) return false;

  // valid
  return true;
};
