// body-parser/lib/types/json.js
/*!
 * body-parser
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var bytes = require('bytes')
var contentType = require('content-type')
var createError = require('http-errors')
var debug = require('debug')('body-parser:json')
var read = require('../read')
var typeis = require('type-is')

/**
 * Module exports.
 */

module.exports = json

/**
 * RegExp to match the first non-space in a string.
 *
 * Allowed whitespace is defined in RFC 7159:
 *
 *    ws = *(   ; whitespace
 *            %x20 |   ; Space
 *            %x09 |   ; Horizontal tab
 *            %x0A |   ; Line feed or New line
 *            %x0D )   ; Carriage return
 *
 * @private
 */

var FIRST_CHAR_REGEXP = /^[\x20\x09\x0a\x0d]*(.)/

/**
 * Create a middleware to parse JSON bodies.
 *
 * @param {object} [options]
 * @return {function}
 * @public
 */

function json(options) {
  var opts = options || {}

  var limit = typeof opts.limit !== 'number'
    ? bytes.parse(opts.limit || '100kb')
    : opts.limit
  var inflate = opts.inflate !== false
  var reviver = opts.reviver
  var strict = opts.strict !== false
  var type = opts.type || 'application/json'
  var verify = opts.verify || false

  if (verify !== false && typeof verify !== 'function') {
    throw new TypeError('option verify must be function')
  }

  // create the appropriate type checking function
  var shouldParse = typeof type !== 'function'
    ? createTypeChecker(type)
    : type

  function parse(body) {
    if (body.length === 0) {
      // special-case empty json body, as it's a common client-side mistake
      // TODO: maybe make this configurable or part of "strict" option
      return {}
    }

    if (strict) {
      var first = firstChar(body)

      if (first !== '{' && first !== '[') {
        debug('strict violation: %s', body)
        throw createStrictSyntaxError(body, first)
      }
    }

    try {
      debug('parse json')
      return JSON.parse(body, reviver)
    } catch (e) {
      debug('parse json failed: %o', e)
      throw createParseError(body, e)
    }
  }

  return function jsonParser(req, res, next) {
    if (req._body) {
      debug('body already parsed')
      next()
      return
    }

    req.body = req.body || {}

    // skip requests without bodies
    if (!typeis.hasBody(req)) {
      debug('skip empty body')
      next()
      return
    }

    debug('content-type %j', req.headers['content-type'])

    // determine if request should be parsed
    if (!shouldParse(req)) {
      debug('skip parsing')
      next()
      return
    }

    // assert charset
    var charset = getCharset(req) || 'utf-8'
    if (charset.substr(0, 4) !== 'utf-') {
      debug('invalid charset')
      next(createError(415, 'unsupported charset "' + charset.toUpperCase() + '"', {
        charset: charset,
        type: 'charset.unsupported'
      }))
      return
    }

    // read
    read(req, res, next, parse, debug, {
      encoding: charset,
      inflate: inflate,
      limit: limit,
      verify: verify
    })
  }
}

/**
 * Create strict violation syntax error matching native error.
 *
 * @param {string} str
 * @param {string} char
 * @return {Error}
 * @private
 */

function createStrictSyntaxError(str, char) {
  var index = str.indexOf(char)
  var partial = str.substring(0, index) + '#'

  try {
    JSON.parse(partial); /* istanbul ignore next */ throw new SyntaxError('strict violation')
  } catch (e) {
    return createError(400, e)
  }
}

/**
 * Create a JSON parse error.
 *
 * @param {string} body
 * @param {Error} err
 * @return {Error}
 * @private
 */

function createParseError(body, err) {
  var error = createError(400, err)

  if (error.expose) {
    error.body = body
  }

  return error
}

/**
 * Create type checking function for given content type.
 *
 * @param {string|array} type
 * @return {function}
 * @private
 */

function createTypeChecker(type) {
  return function checkType(req) {
    // get the content type
    var value = req.headers['content-type']

    // no content type
    if (!value) {
      debug('missing content-type')
      return false
    }

    debug('parse content-type: %s', value)

    // compare content-type
    return typeis.is(value, type) !== false
  }
}

/**
 * Get the first non-whitespace character in a string.
 *
 * @param {string} str
 * @return {string}
 * @private
 */

function firstChar(str) {
  var match = FIRST_CHAR_REGEXP.exec(str)
  return match ? match[1] : ''
}

/**
 * Get the charset of a request.
 *
 * @param {object} req
 * @return {string|undefined}
 * @private
 */

function getCharset(req) {
  try {
    return contentType.parse(req).parameters.charset
  } catch (e) {
    return undefined
  }
}
