// body-parser/lib/types/raw.js
/*!
 * body-parser
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var bytes = require('bytes')
var contentType = require('content-type')
var debug = require('debug')('body-parser:raw')
var read = require('../read')
var typeis = require('type-is')

/**
 * Module exports.
 */

module.exports = raw

/**
 * Create a middleware to parse raw bodies.
 *
 * @param {object} [options]
 * @return {function}
 * @public
 */

function raw(options) {
  var opts = options || {}

  var limit = typeof opts.limit !== 'number'
    ? bytes.parse(opts.limit || '100kb')
    : opts.limit
  var inflate = opts.inflate !== false
  var type = opts.type || 'application/octet-stream'
  var verify = opts.verify || false

  if (verify !== false && typeof verify !== 'function') {
    throw new TypeError('option verify must be function')
  }

  // create the appropriate type checking function
  var shouldParse = typeof type !== 'function'
    ? createTypeChecker(type)
    : type

  function parse(body) {
    debug('parse raw')
    return body
  }

  return function rawParser(req, res, next) {
    if (req._body) {
      debug('body already parsed')
      next()
      return
    }

    req.body = req.body || {}

    // skip requests without bodies
    if (!typeis.hasBody(req)) {
      debug('skip empty body')
      next()
      return
    }

    debug('content-type %j', req.headers['content-type'])

    // determine if request should be parsed
    if (!shouldParse(req)) {
      debug('skip parsing')
      next()
      return
    }

    // read
    read(req, res, next, parse, debug, {
      encoding: null,  // raw bodies are buffers
      inflate: inflate,
      limit: limit,
      verify: verify
    })
  }
}

/**
 * Create type checking function for given content type.
 *
 * @param {string|array} type
 * @return {function}
 * @private
 */

function createTypeChecker(type) {
  return function checkType(req) {
    // get the content type
    var value = req.headers['content-type']

    // no content type
    if (!value) {
      debug('missing content-type')
      return false
    }

    debug('parse content-type: %s', value)

    // compare content-type
    return typeis.is(value, type) !== false
  }
}
