// body-parser/lib/types/urlencoded.js
/*!
 * body-parser
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var bytes = require('bytes')
var contentType = require('content-type')
var createError = require('http-errors')
var debug = require('debug')('body-parser:urlencoded')
var deprecate = require('depd')('body-parser')
var read = require('../read')
var typeis = require('type-is')
var parse = require('../parse')

/**
 * Module exports.
 */

module.exports = urlencoded

/**
 * Cache of parser modules.
 */

var parsers = Object.create(null)

/**
 * Create a middleware to parse urlencoded bodies.
 *
 * @param {object} [options]
 * @return {function}
 * @public
 */

function urlencoded (options) {
  var opts = options || {}

  // notice extended default
  var extended = opts.extended !== false

  // defaults
  var inflate = opts.inflate !== false
  var limit = typeof opts.limit !== 'number'
    ? bytes.parse(opts.limit || '100kb')
    : opts.limit
  var type = opts.type || 'application/x-www-form-urlencoded'
  var verify = opts.verify || false

  if (verify !== false && typeof verify !== 'function') {
    throw new TypeError('option verify must be function')
  }

  // create the appropriate query parser
  var queryparse = extended
    ? extendedParser(opts)
    : simpleParser(opts)

  // create the appropriate type checking function
  var shouldParse = typeof type !== 'function'
    ? typeChecker(type)
    : type

  function parse (body) {
    return body.length
      ? queryparse(body)
      : {}
  }

  return function urlencodedParser (req, res, next) {
    if (req._body) {
      debug('body already parsed')
      next()
      return
    }

    req.body = req.body || {}

    // skip requests without bodies
    if (!typeis.hasBody(req)) {
      debug('skip empty body')
      next()
      return
    }

    debug('content-type %j', req.headers['content-type'])

    // determine if request should be parsed
    if (!shouldParse(req)) {
      debug('skip parsing')
      next()
      return
    }

    // flag as parsed
    req._body = true

    // parse
    debug('parse urlencoded')
    parse.parse(req, res, next, {
      debug: debug,
      encoding: opts.encoding || 'utf-8',
      inflate: inflate,
      limit: limit,
      parameterLimit: opts.parameterLimit,
      type: type,
      verify: verify
    }, parse)
  }
}

/**
 * Get the extended query parser.
 *
 * @param {object} options
 */

function extendedParser (options) {
  var parameterLimit = options.parameterLimit !== undefined
    ? options.parameterLimit
    : 1000
  var parse = parser('qs')

  if (isNaN(parameterLimit) || parameterLimit < 1) {
    throw new TypeError('option parameterLimit must be a positive number')
  }

  if (isFinite(parameterLimit)) {
    parameterLimit = parameterLimit | 0
  }

  return function queryparse (body) {
    var paramCount = 0
    var params = {}

    // make a copy of the body so we can mutate it without affecting the client
    var bodyStr = body.toString()

    // count the number of parameters
    paramCount = (bodyStr.match(/[^=&]+=(?:[^&]*)/g) || []).length

    // check parameters count
    if (paramCount > parameterLimit) {
      throw createError(413, 'too many parameters')
    }

    // parse
    params = parse(body, {
      allowPrototypes: true,
      arrayLimit: Infinity,
      depth: Infinity,
      parameterLimit: parameterLimit
    })

    return params
  }
}

/**
 * Get the simple query parser.
 *
 * @param {object} options
 */

function simpleParser (options) {
  var parameterLimit = options.parameterLimit !== undefined
    ? options.parameterLimit
    : 1000
  var parse = parser('querystring')

  if (isNaN(parameterLimit) || parameterLimit < 1) {
    throw new TypeError('option parameterLimit must be a positive number')
  }

  if (isFinite(parameterLimit)) {
    parameterLimit = parameterLimit | 0
  }

  return function queryparse (body) {
    var paramCount = 0
    var params = {}

    // make a copy of the body so we can mutate it without affecting the client
    var bodyStr = body.toString()

    // count the number of parameters
    paramCount = (bodyStr.match(/[^=&]+=(?:[^&]*)/g) || []).length

    // check parameters count
    if (paramCount > parameterLimit) {
      throw createError(413, 'too many parameters')
    }

    // parse
    params = parse(body, {
      allowPrototypes: true
    })

    return params
  }
}

/**
 * Get parser for module name dynamically.
 *
 * @param {string} name
 * @return {function}
 * @private
 */

function parser (name) {
  var mod = parsers[name]

  if (mod !== undefined) {
    return mod.parse
  }

  // load module
  mod = require(name)

  // store parser
  parsers[name] = mod

  return mod.parse
}

/**
 * Get type checking function for type.
 *
 * @param {string|array} type
 * @return {function}
 */

function typeChecker (type) {
  return function checkType (req) {
    return Boolean(typeis(req, type))
  }
}
