// body-parser/lib/errors.js
/**
 * body-parser error handling module
 * 
 * This module defines custom error types and error creation utilities
 * used across all body-parser implementations to ensure consistent
 * error handling for malformed requests.
 *
 * @module
 * @private
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */
var http = require('http')
var setPrototypeOf = require('setprototypeof')
var statuses = require('statuses')
var inherits = require('inherits')

/**
 * Module exports.
 * @public
 */

module.exports = {
  createError: createError,
  HttpError: createHttpErrorConstructor(),
  ContentTypeError: createContentTypeError(),
  LimitError: createLimitError(),
  ParseError: createParseError(),
  CharsetError: createCharsetError(),
  isBodParserError: createIsBodyParserErrorFunction()
}

/**
 * Create a new HTTP Error for body-parser.
 *
 * @param {string} message Error message.
 * @param {number} status Status code.
 * @param {object} [props] Additional properties to include on the error.
 * @return {Error} The error object.
 * @public
 */
function createError(message, status, props) {
  // Create error
  var error = new Error(message)
  
  // Capture stack trace
  Error.captureStackTrace(error, createError)
  
  // Set status code
  error.status = error.statusCode = status || 400
  
  // Set properties
  if (props) {
    for (var key in props) {
      if (key !== 'status' && key !== 'statusCode') {
        error[key] = props[key]
      }
    }
  }
  
  return error
}

/**
 * Create HTTP error constructor.
 * @private
 */
function createHttpErrorConstructor() {
  function HttpError(message, status) {
    // Create the error object
    var err = new Error(message || statuses[status] || 'Body Parser Error')
    
    // Capture a stack trace
    Error.captureStackTrace(err, this.constructor)
    
    // Set the prototype
    setPrototypeOf(err, this.constructor.prototype)
    
    // Assign status properties
    err.status = err.statusCode = status || 400
    err.expose = true
    
    return err
  }
  
  inherits(HttpError, Error)
  HttpError.prototype.name = 'HttpError'
  
  return HttpError
}

/**
 * Create a constructor for content type errors.
 * @private
 */
function createContentTypeError() {
  var HttpError = module.exports.HttpError
  
  function ContentTypeError(message) {
    var err = new HttpError(message || 'Invalid content type', 415)
    
    // Capture a stack trace
    Error.captureStackTrace(err, this.constructor)
    
    // Set the prototype
    setPrototypeOf(err, this.constructor.prototype)
    
    return err
  }
  
  inherits(ContentTypeError, HttpError)
  ContentTypeError.prototype.name = 'ContentTypeError'
  ContentTypeError.prototype.type = 'entity.parse.content_type'
  
  return ContentTypeError
}

/**
 * Create a constructor for limit exceeded errors.
 * @private
 */
function createLimitError() {
  var HttpError = module.exports.HttpError
  
  function LimitError(message, type, limit) {
    var err = new HttpError(message || 'Request entity too large', 413)
    
    // Capture a stack trace
    Error.captureStackTrace(err, this.constructor)
    
    // Set the prototype
    setPrototypeOf(err, this.constructor.prototype)
    
    // Set limit properties
    err.type = type || 'entity.too.large'
    err.limit = limit
    
    return err
  }
  
  inherits(LimitError, HttpError)
  LimitError.prototype.name = 'LimitError'
  
  return LimitError
}

/**
 * Create a constructor for parse errors.
 * @private
 */
function createParseError() {
  var HttpError = module.exports.HttpError
  
  function ParseError(message, body, type) {
    var err = new HttpError(message || 'Invalid request body', 400)
    
    // Capture a stack trace
    Error.captureStackTrace(err, this.constructor)
    
    // Set the prototype
    setPrototypeOf(err, this.constructor.prototype)
    
    // Set parse error properties
    err.body = body
    err.type = type || 'entity.parse.failed'
    
    return err
  }
  
  inherits(ParseError, HttpError)
  ParseError.prototype.name = 'ParseError'
  
  return ParseError
}

/**
 * Create a constructor for charset errors.
 * @private
 */
function createCharsetError() {
  var HttpError = module.exports.HttpError
  
  function CharsetError(message, charset) {
    var err = new HttpError(message || 'Unsupported charset', 415)
    
    // Capture a stack trace
    Error.captureStackTrace(err, this.constructor)
    
    // Set the prototype
    setPrototypeOf(err, this.constructor.prototype)
    
    // Set charset properties
    err.charset = charset
    err.type = 'charset.unsupported'
    
    return err
  }
  
  inherits(CharsetError, HttpError)
  CharsetError.prototype.name = 'CharsetError'
  
  return CharsetError
}

/**
 * Create isBodyParserError function
 * @private
 */
function createIsBodyParserErrorFunction() {
  var HttpError = module.exports.HttpError
  var ContentTypeError = module.exports.ContentTypeError
  var LimitError = module.exports.LimitError
  var ParseError = module.exports.ParseError
  var CharsetError = module.exports.CharsetError
  
  return function isBodyParserError(val) {
    if (!val || typeof val !== 'object') {
      return false
    }
    
    return val instanceof HttpError ||
           val instanceof ContentTypeError ||
           val instanceof LimitError ||
           val instanceof ParseError ||
           val instanceof CharsetError ||
           (val instanceof Error && 
            typeof val.status === 'number' && 
            typeof val.type === 'string' && 
            val.type.indexOf('entity.') === 0)
  }
}
