// body-parser/lib/parse.js
/**
 * Module dependencies.
 */

var bytes = require('bytes')
var contentType = require('content-type')
var errors = require('./errors')
var debug = require('debug')('body-parser:parse')
var read = require('./read')
var typeis = require('type-is')

/**
 * Module exports.
 */

module.exports = parse

/**
 * RegExp to match non-URL encoded characters in charset value.
 * @private
 */

var CHARSET_REGEXP = /[^\x20-\x7E\x80-\xFF]/

/**
 * RegExp to match non-URL encoded characters in unquoted charset value.
 * @private
 */

var UNQUOTED_CHARSET_REGEXP = /[\x00-\x1F\x7F]/

/**
 * Parse the request body.
 *
 * @param {object} req
 * @param {object} res
 * @param {function} next
 * @param {object} options
 * @param {function} parse
 * @public
 */

function parse (req, res, next, options, parse) {
  // get content-type
  var contentType = req.headers['content-type'] || ''
  var hasBody = req.headers['transfer-encoding'] !== undefined ||
    !isNaN(req.headers['content-length'])

  // determine if request has a body
  if (!hasBody) {
    debug('skip empty body')
    next()
    return
  }

  // determine if request type is parsable
  if (!typeis.hasBody(req)) {
    debug('skip empty body')
    next()
    return
  }

  debug('content-type %j', contentType)

  // get type
  var type = options.type

  // determine if request content-type matches type
  var typeMatch = false
  try {
    typeMatch = typeis(req, type)
  } catch (err) {
    next(new errors.ContentTypeError(err.message))
    return
  }

  // handle invalid content-type
  if (!typeMatch) {
    debug('skip parsing: content-type doesn\'t match %j', type)
    next()
    return
  }

  // get charset
  var charset
  try {
    charset = getCharset(contentType, options) || 'utf-8'
  } catch (err) {
    next(new errors.CharsetError('unsupported charset "' + err.charset + '"', err.charset))
    return
  }

  // get encoding
  var encoding = getEncoding(charset)

  // validate charset
  if (encoding === null) {
    debug('skip parsing: unsupported charset %j', charset)
    next(new errors.CharsetError('unsupported charset "' + charset + '"', charset))
    return
  }

  // get verify function
  var verify = options.verify

  // verify
  if (verify !== undefined && typeof verify !== 'function') {
    next(errors.createError('verify option must be a function', 500, {
      type: 'options.verify.invalid'
    }))
    return
  }

  // get inflate
  var inflate = options.inflate !== false

  // get limit
  var limit = bytes.parse(options.limit)

  // get parameterLimit
  var parameterLimit = options.parameterLimit !== undefined
    ? options.parameterLimit
    : 1000

  // read body
  debug('read body')
  read(req, res, next, parse, debug, {
    encoding: encoding,
    inflate: inflate,
    limit: limit,
    verify: verify
  })
}

/**
 * Get the charset of a request.
 *
 * @param {string} contentType
 * @param {object} [options]
 * @return {string|undefined}
 * @private
 */

function getCharset (contentType, options) {
  if (!contentType) {
    return undefined
  }

  var parsed = contentType.parse(contentType)
  var charset = parsed.parameters.charset

  if (charset) {
    return charset
  }

  // default charset
  if (options.defaultCharset) {
    return options.defaultCharset
  }

  return undefined
}

/**
 * Get the encoding for the charset.
 *
 * @param {string} charset
 * @return {string|null}
 * @private
 */

function getEncoding (charset) {
  if (!charset) {
    return null
  }

  try {
    return contentType.parse(charset).toString().toLowerCase()
  } catch (e) {
    return null
  }
}

/**
 * Get parser for the charset.
 *
 * @param {string} charset
 * @return {function|null}
 * @private
 */

function getParser (charset) {
  var encoding = getEncoding(charset)

  if (encoding === null) {
    return null
  }

  switch (encoding) {
    case 'utf-8':
    case 'utf8':
      return decodeURIComponent
    case 'iso-8859-1':
    case 'latin1':
      return decodeLatin1
    default:
      return null
  }
}

/**
 * Decode a Latin-1 string to a string with percent-encoded characters.
 *
 * @param {string} str
 * @return {string}
 * @private
 */

function decodeLatin1 (str) {
  var encoded = ''

  for (var i = 0; i < str.length; i++) {
    var charCode = str.charCodeAt(i)

    if (charCode <= 0x7F) {
      encoded += str.charAt(i)
      continue
    }

    encoded += '%' + charCode.toString(16).toUpperCase()
  }

  return encoded
}

/**
 * Check if the character is a valid token character.
 *
 * @see https://tools.ietf.org/html/rfc7230#section-3.2.6
 * @param {number} code
 * @return {boolean}
 * @private
 */

function isTokenCharacter (code) {
  return code >= 0x30 && code <= 0x39 || // 0-9
    code >= 0x41 && code <= 0x5A || // A-Z
    code >= 0x61 && code <= 0x7A || // a-z
    code === 0x21 || // !
    code === 0x23 || // #
    code === 0x24 || // $
    code === 0x25 || // %
    code === 0x26 || // &
    code === 0x27 || // '
    code === 0x2A || // *
    code === 0x2B || // +
    code === 0x2D || // -
    code === 0x2E || // .
    code === 0x5E || // ^
    code === 0x5F || // _
    code === 0x60 || // `
    code === 0x7C || // |
    code === 0x7E    // ~
}

/**
 * Validate a charset name.
 *
 * @param {string} charset
 * @return {boolean}
 * @private
 */

function validateCharset (charset) {
  // check for non-URL encoded characters
  if (CHARSET_REGEXP.test(charset)) {
    return false
  }

  // check if charset is a token or quoted-string
  if (charset.charCodeAt(0) === 0x22 /* " */) {
    // remove quotes
    charset = charset.slice(1, -1)

    // check for non-URL encoded characters in quoted value
    if (UNQUOTED_CHARSET_REGEXP.test(charset)) {
      return false
    }

    // ensure charset is a token
    for (var i = 0; i < charset.length; i++) {
      if (!isTokenCharacter(charset.charCodeAt(i))) {
        return false
      }
    }
  }

  return true
}
