// body-parser/lib/read.js
/*!
 * body-parser
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var createError = require('http-errors')
var destroy = require('destroy')
var getBody = require('raw-body')
var iconv = require('iconv-lite')
var onFinished = require('on-finished')
var zlib = require('zlib')

/**
 * Module exports.
 */

module.exports = read

/**
 * Read a request into a buffer and parse.
 *
 * @param {object} req
 * @param {object} res
 * @param {function} next
 * @param {function} parse
 * @param {function} debug
 * @param {object} options
 * @public
 */

function read (req, res, next, parse, debug, options) {
  var length = req.headers['content-length']
  var encoding = req.headers['content-encoding'] || 'identity'
  var inflate = options.inflate !== false
  var limit = options.limit
  var type = options.type || ''
  var verify = options.verify

  // Skip requests without bodies
  if (!typeIs.hasBody(req)) {
    debug('skip empty body')
    return next()
  }

  // Skip unsupported content encoding
  if (inflate === false && encoding !== 'identity') {
    debug('skip unsupported content encoding "%s"', encoding)
    return next(createError(415, 'unsupported content encoding "' + encoding + '"', {
      encoding: encoding,
      type: 'encoding.unsupported'
    }))
  }

  // Check content type
  if (type !== '' && !typeIs(req, type)) {
    debug('skip unsupported content type')
    return next()
  }

  // Request body too large
  if (length !== undefined && limit !== undefined && bytesLength(length) > bytesLength(limit)) {
    debug('request entity too large')
    return next(createError(413, 'request entity too large', {
      expected: bytesLength(limit),
      length: bytesLength(length),
      limit: limit,
      type: 'entity.too.large'
    }))
  }

  // Handle request body
  readBody(req, res, next, parse, debug, options, encoding, inflate, limit, verify)
}

/**
 * Get the content stream of the request.
 *
 * @param {object} req
 * @param {string} encoding
 * @param {boolean} inflate
 * @return {object}
 * @private
 */

function getStream (req, encoding, inflate) {
  var stream = req

  // Handle compressed bodies
  if (inflate && encoding !== 'identity') {
    switch (encoding) {
      case 'deflate':
        stream = stream.pipe(zlib.createInflate())
        break
      case 'gzip':
        stream = stream.pipe(zlib.createGunzip())
        break
    }
  }

  return stream
}

/**
 * Read the body of the request.
 *
 * @param {object} req
 * @param {object} res
 * @param {function} next
 * @param {function} parse
 * @param {function} debug
 * @param {object} options
 * @param {string} encoding
 * @param {boolean} inflate
 * @param {string} limit
 * @param {function} verify
 * @private
 */

function readBody (req, res, next, parse, debug, options, encoding, inflate, limit, verify) {
  var contentLength = req.headers['content-length']
  var stream = getStream(req, encoding, inflate)
  var received = 0
  var decoder

  // Set up decoder for unzipped streams
  if (options.encoding !== null) {
    decoder = iconv.getDecoder(options.encoding)
  }

  // Handle stream errors
  stream.on('error', function onStreamError (err) {
    // If the request aborted, then stop reading
    if (err.code === 'ECONNABORTED') {
      received = -1
      stream.destroy()
      return
    }

    // If the request errored, forward the error
    next(createError(400, err, {
      code: err.code,
      expected: bytesLength(limit),
      length: bytesLength(contentLength),
      received: received,
      type: 'stream.error'
    }))

    // Unpipe from stream and destroy
    stream.unpipe()
    stream.destroy()
  })

  // Get body
  getBody(stream, {
    encoding: options.encoding === null ? null : 'utf8',
    length: contentLength,
    limit: limit
  }, function (err, body) {
    if (err) {
      // Default to 400 for errors
      var status = err.status || 400
      var type = err.type

      // Read error occurred
      return next(createError(status, err, {
        code: err.code,
        expected: bytesLength(limit),
        length: bytesLength(contentLength),
        received: received,
        type: type || 'entity.parse.failed'
      }))
    }

    // Verify
    if (verify) {
      try {
        verify(req, res, body, encoding)
      } catch (err) {
        return next(createError(403, err, {
          body: body,
          type: 'entity.verify.failed'
        }))
      }
    }

    // Parse
    var str = body
    try {
      // Decode strings if needed
      if (typeof body !== 'string' && decoder) {
        str = decoder.write(body) + decoder.end()
      }

      // Parse body
      var parsed = parse(str)
      debug('parse %s', str)

      // Set body for next middleware
      req.body = parsed
      next()
    } catch (err) {
      // Parse error occurred
      return next(createError(400, err, {
        body: str,
        type: 'entity.parse.failed'
      }))
    }
  })

  // Listen for request aborted
  onFinished(req, function onRequestAborted () {
    if (received !== -1) {
      // Destroy the stream
      stream.unpipe()
      stream.destroy()
    }
  })
}

/**
 * Get the byte length of a string or buffer.
 *
 * @param {string|number|object} val
 * @return {number}
 * @private
 */

function bytesLength (val) {
  if (val === undefined) {
    return undefined
  }

  if (typeof val === 'string') {
    return Buffer.byteLength(val)
  }

  if (typeof val === 'number') {
    return val
  }

  // Buffer/ArrayBuffer
  if (val && typeof val.byteLength === 'number') {
    return val.byteLength
  }

  return undefined
}

/**
 * Get the simple type checker.
 *
 * @param {object} req
 * @param {string|array} types
 * @return {boolean}
 * @private
 */

function typeIs (req, types) {
  var typeis = require('type-is')
  return typeis(req, types)
}

// For backwards compatibility
typeIs.hasBody = function hasBody (req) {
  var headers = req.headers
  var contentLength = headers['content-length']

  if (contentLength === undefined || contentLength === '') {
    return false
  }

  var length = parseInt(contentLength, 10)
  return isNaN(length) === false && length > 0
}
