// range-parser/index.js
/*!
 * range-parser
 * Copyright(c) 2012-2014 TJ Holowaychuk
 * Copyright(c) 2015-2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module exports.
 * @public
 */

module.exports = rangeParser

/**
 * Parse "Range" header `str` relative to the given file `size`.
 *
 * @param {Number} size
 * @param {String} str
 * @param {Object} [options]
 * @return {Array|Number} Ranges or negative number indicating error
 * @public
 */

function rangeParser (size, str, options) {
  if (typeof str !== 'string') {
    return -2
  }

  // Default options
  var opts = options || {}

  // Parse ranges
  var arr = str.slice(str.indexOf('=') + 1).split(',')
  var ranges = []

  // If no ranges, return error
  if (arr.length === 0) {
    return -1
  }

  // Parse all ranges
  for (var i = 0; i < arr.length; i++) {
    var range = parseRange(arr[i], size)
    
    // Handle unsatisfiable range
    if (range === -1) {
      return -1
    }
    
    // Handle invalid range
    if (range === -2) {
      // Continue to next range
      continue
    }
    
    // Add range
    ranges.push(range)
  }

  // If no ranges, return error
  if (ranges.length === 0) {
    return -1
  }

  return opts.combine !== false
    ? combineRanges(ranges)
    : ranges
}

/**
 * Combine overlapping & adjacent ranges.
 * @private
 */

function combineRanges (ranges) {
  var ordered = ranges.map(mapWithIndex).sort(sortByRangeStart)
  var result = ordered.reduce(function (result, range, i, arr) {
    // Skip first item
    if (i === 0) {
      return [range]
    }

    // Get previous range
    var prev = result[result.length - 1]

    // Check if adjacent or overlapping
    if (prev.end + 1 >= range.start) {
      // Merge ranges
      prev.end = Math.max(prev.end, range.end)
      return result
    }

    // Add to result
    result.push(range)
    return result
  }, [])

  // Map back to original structure
  return result.map(function (range) {
    return {
      start: range.start,
      end: range.end
    }
  })
}

/**
 * Map array item with index.
 * @private
 */

function mapWithIndex (range, index) {
  return {
    start: range.start,
    end: range.end,
    index: index
  }
}

/**
 * Parse a range header string into a range object.
 * @private
 */

function parseRange (str, size) {
  var valid = true
  var i = str.indexOf('-')
  var start, end

  if (i === -1) {
    return -2
  }

  // Extract start & end positions
  start = str.substring(0, i)
  end = str.substring(i + 1)

  // Handle missing start
  if (start === '') {
    start = size - parseInt(end, 10)
    end = size - 1

    // Handle invalid negative value for start
    if (start < 0) start = 0
  } else {
    // Parse start value
    start = parseInt(start, 10)

    // Handle missing end
    if (end === '') {
      end = size - 1
    } else {
      // Parse end value
      end = parseInt(end, 10)
    }
  }

  // Check for invalid values
  // - start or end positions are non-numbers
  // - start position is greater than end position
  // - start position is greater than file size
  if (isNaN(start) || isNaN(end) || start > end || start >= size) {
    return -1
  }

  // Limit end to file size - 1
  if (end >= size) end = size - 1

  return {
    start: start,
    end: end
  }
}

/**
 * Sort function for range array start position.
 * @private
 */

function sortByRangeStart (a, b) {
  return a.start - b.start
}
