// proxy-addr/index.js
/**
 * Module dependencies.
 */

var forwarded = require('forwarded')
var ipaddr = require('ipaddr.js')

/**
 * Module exports.
 */

module.exports = proxyaddr
module.exports.all = alladdrs
module.exports.compile = compile

/**
 * Variables.
 */

var DIGIT_REGEXP = /^\d+$/

/**
 * Parse the X-Forwarded-For header.
 *
 * @param {object} req
 * @param {function|array|string} trust
 * @public
 */

function proxyaddr(req, trust) {
  if (!req) {
    throw new TypeError('req argument is required')
  }

  if (!trust) {
    throw new TypeError('trust argument is required')
  }

  var addrs = alladdrs(req)
  var addr = addrs[0]
  var trusted = compile(trust)

  if (!Array.isArray(addrs) || addrs.length === 0) {
    return addr
  }

  // Find the first untrusted address
  for (var i = 1; i < addrs.length; i++) {
    if (!trusted(addrs[i], i - 1)) {
      break
    }
    addr = addrs[i]
  }

  return addr
}

/**
 * Parse X-Forwarded-For header for all addresses.
 *
 * @param {object} req
 * @return {array}
 * @public
 */

function alladdrs(req) {
  if (!req) {
    throw new TypeError('req argument is required')
  }

  // Get all addresses
  var addrs = forwarded(req)

  if (!Array.isArray(addrs) || addrs.length === 0) {
    return [req.connection.remoteAddress || '127.0.0.1']
  }

  // Return all addresses
  return addrs
}

/**
 * Compile argument into trust function.
 *
 * @param {function|array|string} val
 * @return {function}
 * @private
 */

function compile(val) {
  if (typeof val === 'function') {
    return val
  }

  if (val === true) {
    // Trust all proxies
    return function() { return true }
  }

  if (typeof val === 'number') {
    // Trust the first n proxies
    return function(addr, i) { return i < val }
  }

  if (typeof val === 'string') {
    // Trust a single IP address or CIDR
    return compileTrust(val.split(',').map(function(v) { return v.trim() }))
  }

  // Trust an array of IP addresses or CIDRs
  return compileTrust(val)
}

/**
 * Compile array of trust values into trust function.
 *
 * @param {array} arr
 * @return {function}
 * @private
 */

function compileTrust(arr) {
  if (!Array.isArray(arr)) {
    throw new TypeError('argument is not an array')
  }

  var trusts = arr.map(function(val) {
    if (typeof val !== 'string') {
      throw new TypeError('trust value must be a string')
    }

    // IPv4 CIDR or exact match
    if (val.indexOf('/') !== -1) {
      return compileCIDR(val)
    }

    // IPv6 CIDR or exact match
    if (val.indexOf(':') !== -1) {
      return compileIPv6(val)
    }

    // Wildcard match
    if (val === '*') {
      return function() { return true }
    }

    // IPv4 exact match
    return compileIPv4(val)
  })

  return function(addr) {
    if (!addr) return false

    var ip
    try {
      ip = ipaddr.parse(addr)
    } catch (e) {
      return false
    }

    var trusted = false
    for (var i = 0; i < trusts.length; i++) {
      if (trusts[i](ip)) {
        trusted = true
        break
      }
    }

    return trusted
  }
}

/**
 * Compile IPv4 CIDR into trust function.
 *
 * @param {string} cidr
 * @return {function}
 * @private
 */

function compileCIDR(cidr) {
  var parts = cidr.split('/')
  var subnetMask = parts[1]

  if (!subnetMask || !DIGIT_REGEXP.test(subnetMask)) {
    throw new TypeError('invalid CIDR subnet mask')
  }

  var subnet = ipaddr.parseCIDR(cidr)

  return function(addr) {
    return addr.kind() === 'ipv4' && addr.match(subnet)
  }
}

/**
 * Compile IPv4 address into trust function.
 *
 * @param {string} addr
 * @return {function}
 * @private
 */

function compileIPv4(addr) {
  var ip = parseIPv4(addr)

  return function(other) {
    return other.kind() === 'ipv4' && other.toString() === ip.toString()
  }
}

/**
 * Compile IPv6 address into trust function.
 *
 * @param {string} addr
 * @return {function}
 * @private
 */

function compileIPv6(addr) {
  if (addr.indexOf('/') !== -1) {
    // IPv6 CIDR
    var subnet = ipaddr.parseCIDR(addr)
    return function(addr) {
      return addr.kind() === 'ipv6' && addr.match(subnet)
    }
  }

  // IPv6 exact match
  var ip = parseIPv6(addr)
  return function(other) {
    return other.kind() === 'ipv6' && other.toString() === ip.toString()
  }
}

/**
 * Parse IPv4 address string into ipaddr.js object.
 *
 * @param {string} addr
 * @return {object}
 * @private
 */

function parseIPv4(addr) {
  try {
    var ip = ipaddr.parse(addr)

    if (ip.kind() !== 'ipv4') {
      throw new Error('not ipv4')
    }

    return ip
  } catch (e) {
    throw new TypeError('invalid IPv4 address')
  }
}

/**
 * Parse IPv6 address string into ipaddr.js object.
 *
 * @param {string} addr
 * @return {object}
 * @private
 */

function parseIPv6(addr) {
  try {
    var ip = ipaddr.parse(addr)

    if (ip.kind() !== 'ipv6') {
      throw new Error('not ipv6')
    }

    return ip
  } catch (e) {
    throw new TypeError('invalid IPv6 address')
  }
}
