// proxy-addr/index.js
/*!
 * proxy-addr
 * Copyright(c) 2014-2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var forwarded = require('forwarded')
var ipaddr = require('ipaddr.js')

/**
 * Module exports.
 * @public
 */

module.exports = proxyaddr
module.exports.all = alladdrs
module.exports.compile = compile

/**
 * Variables.
 * @private
 */

var DIGIT_REGEXP = /^\d+$/
var IP_REGEXP = /^[0-9a-f:.]+$/i

/**
 * Get all addresses in the request, optionally filtered by trust function.
 *
 * @param {Object} request
 * @param {Function|Array|String} trust
 * @public
 */

function alladdrs(req, trust) {
  // Get addresses
  var addrs = forwarded(req)

  if (!trust) {
    // Return all addresses
    return addrs
  }

  if (typeof trust !== 'function') {
    trust = compile(trust)
  }

  for (var i = 0; i < addrs.length - 1; i++) {
    if (trust(addrs[i], i)) continue

    addrs.splice(0, i + 1)
    break
  }

  return addrs
}

/**
 * Compile argument into trust function.
 *
 * @param {Array|String} val
 * @private
 */

function compile(val) {
  if (!val) {
    throw new TypeError('argument is required')
  }

  var trust

  if (typeof val === 'string') {
    trust = [val]
  } else if (Array.isArray(val)) {
    trust = val.slice()
  } else {
    throw new TypeError('unsupported trust argument')
  }

  for (var i = 0; i < trust.length; i++) {
    val = trust[i]

    if (!Object.prototype.hasOwnProperty.call(val, 'length')) {
      continue
    }

    // Validate CIDR
    var cidr = typeof val === 'string'
      ? val.indexOf('/')
      : -1

    if (cidr !== -1) {
      var subnet = val.substring(0, cidr)
      var bits = val.substring(cidr + 1)

      if (IP_REGEXP.test(subnet) && DIGIT_REGEXP.test(bits)) {
        continue
      }
    }

    if (IP_REGEXP.test(val)) {
      continue
    }

    throw new TypeError('invalid IP address: ' + val)
  }

  return compileTrust(trust)
}

/**
 * Compile `arr` elements into trust function.
 *
 * @param {Array} arr
 * @private
 */

function compileTrust(arr) {
  var subnets = new Array(arr.length)

  for (var i = 0; i < arr.length; i++) {
    var val = arr[i]

    if (IP_REGEXP.test(val)) {
      // IP address
      subnets[i] = parseIPNotation(val)
      continue
    }

    // CIDR notation
    var cidr = val.indexOf('/')
    var subnet = val.substring(0, cidr)
    var bits = val.substring(cidr + 1)

    subnets[i] = parseIPNotation(subnet, bits)
  }

  return trustSubnet

  function trustSubnet(addr) {
    var ip = parseIP(addr)
    if (!ip) return false

    for (var i = 0; i < subnets.length; i++) {
      var subnet = subnets[i]
      var subnetip = subnet[0]
      var subnetbits = subnet[1]
      var familycheck = subnetip.kind() === ip.kind()
      var matching = familycheck && ip.match(subnetip, subnetbits)

      if (matching) return true
    }

    return false
  }
}

/**
 * Parse IP notation string into IPAddr object.
 *
 * @param {String} note
 * @param {String} bits
 * @private
 */

function parseIPNotation(note, bits) {
  var ip = parseIP(note)
  var kind = ip.kind()

  return [ip, bits === undefined
    ? kind === 'ipv4' ? 32 : 128
    : parseInt(bits, 10)]
}

/**
 * Parse IP string into IPAddr object.
 *
 * @param {String} str
 * @private
 */

function parseIP(str) {
  try {
    return ipaddr.parse(str)
  } catch (e) {
    return null
  }
}

/**
 * Get the request's IP address.
 *
 * @param {Object} request
 * @param {Function|Array|String} trust
 * @public
 */

function proxyaddr(req, trust) {
  if (!req) {
    throw new TypeError('req argument is required')
  }

  if (!trust) {
    // Return remote address
    return req.connection.remoteAddress
  }

  // Get all addresses
  var addrs = alladdrs(req, trust)

  // Return the best available address
  return addrs[addrs.length - 1]
}
