// debug/node.js
/**
 * node.js - Node.js-specific implementation of the debug module
 * 
 * This file provides the Node.js-specific implementation of the debug module,
 * which is a dependency of Express.js used for internal logging and debugging.
 * 
 * It handles environment detection, console output formatting, and provides
 * the core debugging functionality for server environments.
 * 
 * Copyright (c) 2014-2017 TJ Holowaychuk <tj@vision-media.ca>
 * Copyright (c) 2018-2021 Josh Junon <josh.junon@protonmail.com>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the 'Software'), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Module dependencies.
 */
const tty = require('tty');
const util = require('util');
const fs = require('fs');
const path = require('path');

/**
 * Detect if we're running in a terminal.
 */
const isatty = tty.isatty(process.stderr.fd);

/**
 * Map of ANSI color codes to their corresponding terminal color codes.
 */
const colors = [
  6, 2, 3, 4, 5, 1 // cyan, green, yellow, blue, magenta, red
];

/**
 * Currently only WebKit-based Web Inspectors, Firefox >= v31,
 * and the Firebug extension (any Firefox version) are known
 * to support "%c" CSS customizations.
 */
const formatters = {
  /**
   * Map %o to util.inspect with options colored output.
   */
  o: function(v) {
    return util.inspect(v, { colors: this.useColors });
  },
  
  /**
   * Map %O to util.inspect without options colored output.
   */
  O: function(v) {
    return util.inspect(v);
  }
};

/**
 * Select a color for a debug namespace.
 * @param {String} namespace The namespace string for the debug instance.
 * @return {Number} An ANSI color code for the given namespace.
 * @api private
 */
function selectColor(namespace) {
  let hash = 0;
  
  for (let i = 0; i < namespace.length; i++) {
    hash = ((hash << 5) - hash) + namespace.charCodeAt(i);
    hash |= 0; // Convert to 32bit integer
  }
  
  return colors[Math.abs(hash) % colors.length];
}

/**
 * Create a debugger with the given `namespace`.
 *
 * @param {String} namespace
 * @return {Function}
 * @api public
 */
function createDebug(namespace) {
  // Define the enabled state based on environment variables
  let enabled;
  
  function debug(...args) {
    // If debugging is not enabled for this namespace, skip
    if (!debug.enabled) return;
    
    const self = debug;
    
    // Set the timestamp
    const curr = new Date();
    const ms = curr - (self.prev || curr);
    self.prev = curr;
    
    // Format the message
    args[0] = coerceToString(args[0]);
    
    if (typeof args[0] !== 'string') {
      // Anything else let's inspect with %O
      args.unshift('%O');
    }
    
    // Apply any formatters
    let index = 0;
    args[0] = args[0].replace(/%([a-zA-Z%])/g, (match, format) => {
      // If we encounter an escaped % then don't increase the array index
      if (match === '%%') return match;
      index++;
      const formatter = formatters[format];
      if (typeof formatter === 'function') {
        const val = args[index];
        match = formatter.call(self, val);
        
        // Now we need to remove this argument from the args array and adjust the index
        args.splice(index, 1);
        index--;
      }
      return match;
    });
    
    // Apply colors to the namespace
    const colorCode = '\u001B[3' + selectColor(namespace) + 'm';
    const prefix = colorCode + namespace + ' ' + '\u001B[0m';
    
    // Format the output with timestamp and namespace
    args[0] = prefix + args[0] + ' +' + ms + 'ms';
    
    // Output to stderr
    process.stderr.write(util.format.apply(util, args) + '\n');
  }
  
  debug.namespace = namespace;
  debug.enabled = isEnabled(namespace);
  debug.useColors = isatty;
  debug.color = selectColor(namespace);
  debug.destroy = destroy;
  debug.extend = extend;
  
  // Env-specific initialization logic for debug instances
  if ('init' in debug) {
    debug.init();
  }
  
  return debug;
}

/**
 * Enables a debug mode by namespaces. This can include modes
 * separated by a colon and wildcards.
 *
 * @param {String} namespaces
 * @api public
 */
function enable(namespaces) {
  createDebug.save(namespaces);
  
  createDebug.names = [];
  createDebug.skips = [];
  
  let i;
  const split = (typeof namespaces === 'string' ? namespaces : '').split(/[\s,]+/);
  const len = split.length;
  
  for (i = 0; i < len; i++) {
    if (!split[i]) continue; // ignore empty strings
    namespaces = split[i].replace(/\*/g, '.*?');
    if (namespaces[0] === '-') {
      createDebug.skips.push(new RegExp('^' + namespaces.substr(1) + '$'));
    } else {
      createDebug.names.push(new RegExp('^' + namespaces + '$'));
    }
  }
}

/**
 * Disable debug output.
 *
 * @return {String} namespaces
 * @api public
 */
function disable() {
  const namespaces = [
    ...createDebug.names.map(toNamespace),
    ...createDebug.skips.map(toNamespace).map(namespace => '-' + namespace)
  ].join(',');
  createDebug.enable('');
  return namespaces;
}

/**
 * Convert regexp to namespace
 *
 * @param {RegExp} regxep
 * @return {String} namespace
 * @api private
 */
function toNamespace(regexp) {
  return regexp.toString()
    .substring(2, regexp.toString().length - 2)
    .replace(/\\\.\*\?/g, '*');
}

/**
 * Coerce `val` to a string.
 *
 * @param {Mixed} val
 * @return {String}
 * @api private
 */
function coerceToString(val) {
  if (val instanceof Error) {
    return val.stack || val.message;
  }
  return val;
}

/**
 * Check if the given namespace is enabled.
 *
 * @param {String} namespace
 * @return {Boolean}
 * @api public
 */
function isEnabled(namespace) {
  if (namespace[namespace.length - 1] === '*') {
    return true;
  }
  
  let i, len;
  
  for (i = 0, len = createDebug.skips.length; i < len; i++) {
    if (createDebug.skips[i].test(namespace)) {
      return false;
    }
  }
  
  for (i = 0, len = createDebug.names.length; i < len; i++) {
    if (createDebug.names[i].test(namespace)) {
      return true;
    }
  }
  
  return false;
}

/**
 * Destroy function for cleanup.
 * @api public
 */
function destroy() {
  console.warn('Instance method `debug.destroy()` is deprecated and no longer does anything. It will be removed in the next major version of `debug`.');
}

/**
 * Extend a namespace with another namespace.
 *
 * @param {String} namespace
 * @return {Function} The extended debug function
 * @api public
 */
function extend(namespace) {
  const delimiter = namespace.includes(':') ? ':' : ':';
  const newDebug = createDebug(this.namespace + delimiter + namespace);
  newDebug.log = this.log;
  return newDebug;
}

/**
 * Load `namespaces`.
 *
 * @return {String} returns the previously persisted debug modes
 * @api private
 */
function load() {
  return process.env.DEBUG || '';
}

/**
 * Save `namespaces`.
 *
 * @param {String} namespaces
 * @api private
 */
function save(namespaces) {
  if (namespaces) {
    process.env.DEBUG = namespaces;
  } else {
    // If you set a process.env field to null or undefined, it gets cast to the
    // string 'null' or 'undefined'. Just delete instead.
    delete process.env.DEBUG;
  }
}

/**
 * Initialize the debug module based on environment variables.
 */
createDebug.enable(load());

/**
 * Expose methods.
 */
createDebug.names = [];
createDebug.skips = [];
createDebug.formatters = formatters;
createDebug.enable = enable;
createDebug.disable = disable;
createDebug.save = save;
createDebug.load = load;

/**
 * Export the debug module.
 */
module.exports = createDebug;

/**
 * Export specific debug namespaces commonly used by Express.js.
 */
module.exports.debug = createDebug('debug');
module.exports.default = createDebug;
module.exports.coerce = coerceToString;
module.exports.disable = disable;
module.exports.enable = enable;
module.exports.enabled = isEnabled;
module.exports.humanize = require('./ms');

/**
 * The currently active debug mode.
 */
module.exports.names = [];
module.exports.skips = [];

/**
 * Map of special "%n" handling functions, for the debug "format" argument.
 *
 * Valid key names are a single, lower or upper-case letter, i.e. "n" and "N".
 */
module.exports.formatters = formatters;
