// depd/index.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var callSiteToString = require('./lib/caller').callSiteToString
var eventListenerCount = require('./lib/compat').eventListenerCount
var relative = require('path').relative

/**
 * Module exports.
 * @public
 */

module.exports = depd

/**
 * Get the path to base files on.
 */

var basePath = process.cwd()

/**
 * Determine if namespace is separated.
 */

function containsNamespace(str) {
  var i = str.indexOf('::')
  return i !== -1 && i < str.length - 2
}

/**
 * Create a deprecation function.
 *
 * @param {string} namespace
 * @param {object} [options]
 * @param {boolean} [options.stream=process.stderr]
 * @param {string} [options.header]
 * @return {function}
 * @public
 */

function depd(namespace, options) {
  if (!namespace) {
    throw new TypeError('argument namespace is required')
  }

  if (typeof namespace !== 'string') {
    throw new TypeError('argument namespace must be a string')
  }

  var opts = options || {}

  // Get environment-specific implementation
  var impl = process.env.NODE_ENV === 'production' || process.env.NO_DEPRECATION
    ? noop
    : process.browser
      ? require('./lib/browser')
      : require('./lib/node')

  // Return a function that will emit deprecation messages
  return function deprecate(message) {
    message = message || 'deprecated'

    if (containsNamespace(message)) {
      message = message.replace(/::/, ' ')
    } else {
      message = namespace + ' ' + message
    }

    return impl(message, opts)
  }
}

/**
 * Display deprecation message.
 * @param {string} message
 * @param {object} options
 * @private
 */

function displayDeprecation(message, options) {
  var opts = options || {}
  var header = opts.header || defaultHeader

  // Format the message
  var formattedMessage = header + ' ' + message

  // Print to the appropriate output stream
  var stream = opts.stream || process.stderr
  var write = stream.isTTY
    ? formatColor(stream, formattedMessage)
    : formatPlain(formattedMessage)

  // Write to the stream
  stream.write(write)
}

/**
 * Format deprecation message without color.
 * @param {string} message
 * @private
 */

function formatPlain(message) {
  return message + '
'
}

/**
 * Format deprecation message with color.
 * @param {object} stream
 * @param {string} message
 * @private
 */

function formatColor(stream, message) {
  var color = stream.isTTY
    ? '\x1b[31;1m'
    : ''
  var reset = stream.isTTY
    ? '\x1b[0m'
    : ''

  return color + message + reset + '\n'
}

/**
 * Get call site location as string.
 * @private
 */

function getStack() {
  var limit = Error.stackTraceLimit
  var obj = {}
  var prep = Error.prepareStackTrace

  Error.prepareStackTrace = prepareObjectStackTrace
  Error.stackTraceLimit = Math.max(10, limit)

  // Capture the stack
  Error.captureStackTrace(obj)

  // Slice this function off the top
  var stack = obj.stack.slice(1)

  Error.prepareStackTrace = prep
  Error.stackTraceLimit = limit

  return stack
}

/**
 * Capture call site stack from v8.
 * @param {object} obj
 * @param {object} stack
 * @private
 */

function prepareObjectStackTrace(obj, stack) {
  return stack
}

/**
 * Return a wrapped function in a deprecation message.
 * @param {Function} fn
 * @param {string} message
 * @private
 */

function wrapfunction(fn, message) {
  if (typeof fn !== 'function') {
    throw new TypeError('argument fn must be a function')
  }

  var args = new Array(arguments.length - 2)

  for (var i = 0; i < args.length; i++) {
    args[i] = arguments[i + 2]
  }

  var deprecatedfn = function () {
    displayDeprecation(message)
    return fn.apply(this, arguments)
  }

  // Copy properties from original function
  for (var key in fn) {
    if (Object.prototype.hasOwnProperty.call(fn, key)) {
      deprecatedfn[key] = fn[key]
    }
  }

  deprecatedfn.name = fn.name
  deprecatedfn.toString = function () {
    return fn.toString()
  }

  return deprecatedfn
}

/**
 * Default deprecation header.
 */

var defaultHeader = 'DeprecationWarning:'

/**
 * No-op function for production environments.
 * @private
 */

function noop() {}
