// debug/src/node.js
/**
 * Module dependencies.
 */

var tty = require('tty');
var util = require('util');
var fs = require('fs');
var path = require('path');

/**
 * Import the common functionality from common.js
 */
var common = require('./common');

/**
 * Expose the module.
 */
var exports = module.exports = common.setup;

/**
 * Copied over from common since we need to use it here directly.
 */
var formatters = exports.formatters = common.formatters;

/**
 * Map %o to util.inspect() with default options.
 */
formatters.o = function(v) {
  this.inspectOpts.colors = this.useColors;
  return util.inspect(v, this.inspectOpts)
    .replace(/\s*\n\s*/g, ' ');
};

/**
 * Map %O to util.inspect() with all options enabled.
 */
formatters.O = function(v) {
  this.inspectOpts.colors = this.useColors;
  return util.inspect(v, this.inspectOpts);
};

/**
 * Check if the terminal supports colors.
 *
 * @return {Boolean}
 * @api public
 */
exports.useColors = function() {
  // Check if stdout is a TTY
  if (process.stdout && !process.stdout.isTTY) {
    return false;
  }

  // Check if stderr is a TTY
  if (process.stderr && !process.stderr.isTTY) {
    return false;
  }

  // Check for Windows
  if (process.platform === 'win32') {
    return true;
  }

  // Check for color support in various environments
  if ('COLORTERM' in process.env) {
    return true;
  }

  if (process.env.TERM === 'dumb') {
    return false;
  }

  if (/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(process.env.TERM)) {
    return true;
  }

  return false;
};

/**
 * Colors for Node.js debug output in TTY mode.
 */
exports.colors = [
  6, 2, 3, 4, 5, 1 // blue, green, yellow, red, magenta, cyan
];

/**
 * Build up the default `inspectOpts` object from the environment variables.
 *
 * @return {Object} The inspectOpts object with defaults applied.
 */
exports.inspectOpts = Object.keys(process.env)
  .filter(function(key) {
    return /^debug_/i.test(key);
  })
  .reduce(function(obj, key) {
    // camel-case
    var prop = key
      .substring(6)
      .toLowerCase()
      .replace(/_([a-z])/g, function(_, k) {
        return k.toUpperCase();
      });

    // coerce string value into JS value
    var val = process.env[key];
    if (/^(yes|on|true|enabled)$/i.test(val)) val = true;
    else if (/^(no|off|false|disabled)$/i.test(val)) val = false;
    else if (val === 'null') val = null;
    else val = Number(val);

    obj[prop] = val;
    return obj;
  }, {});

/**
 * The file descriptor to write to, defaults to stderr.
 */
var fd = parseInt(process.env.DEBUG_FD, 10) || 2;

/**
 * Is stderr a TTY? Colored output is enabled when `true`.
 */
var stderrIsTTY = tty.isatty(fd);

/**
 * Load the `DEBUG` environment variable.
 */
exports.load = common.load;

/**
 * Initialize a debug instance with the given namespace.
 */
exports.init = function(debug) {
  debug.inspectOpts = {};

  var keys = Object.keys(exports.inspectOpts);
  for (var i = 0; i < keys.length; i++) {
    debug.inspectOpts[keys[i]] = exports.inspectOpts[keys[i]];
  }
};

/**
 * Enable namespaces listed in the `DEBUG` environment variable.
 */
exports.enable(exports.load());

/**
 * Output stream, defaulting to stderr.
 */
var stream;

if (process.env.DEBUG_STREAM) {
  try {
    stream = fs.createWriteStream(process.env.DEBUG_STREAM);
  } catch (err) {
    // Handle error gracefully
    console.error('Failed to create debug stream:', err);
    stream = null;
  }
}

if (!stream) {
  stream = fd === 1 ? process.stdout :
           fd === 2 ? process.stderr :
           createWritableStdioStream(fd);
}

/**
 * Create a writable stdio stream from a file descriptor.
 * @param {Number} fd File descriptor
 * @return {stream.Writable} Writable stream
 */
function createWritableStdioStream(fd) {
  var stream;
  try {
    stream = fs.createWriteStream(null, { fd: fd });
    // Make sure the stream is usable
    stream._isStdio = true;
    stream.on('error', function() {
      // Don't throw if the stream has errors, just use process.stderr
      stream = process.stderr;
    });
  } catch (e) {
    // Default to stderr if there's any error
    stream = process.stderr;
  }
  return stream;
}

/**
 * Invokes `util.format()` with the specified arguments and writes to the stream.
 */
exports.log = function() {
  var args = Array.prototype.slice.call(arguments);
  var msg = util.format.apply(util, args);

  stream.write(msg + '\n');
};

/**
 * Save `namespaces`.
 *
 * @param {String} namespaces
 * @api private
 */
exports.save = function(namespaces) {
  if (namespaces) {
    process.env.DEBUG = namespaces;
  } else {
    // If you set a process.env field to null or undefined, it gets cast to the
    // string 'null' or 'undefined'. Just delete instead.
    delete process.env.DEBUG;
  }
};

/**
 * Format args with Node.js TTY colors.
 */
exports.formatArgs = function(args) {
  var name = this.namespace;
  var useColors = this.useColors;

  if (useColors) {
    var c = this.color;
    var colorCode = '\u001b[3' + (c < 8 ? c : '8;5;' + c);
    var prefix = '  ' + colorCode + ';1m' + name + ' ' + '\u001b[0m';

    args[0] = prefix + args[0].split('\n').join('\n' + prefix);
    args.push(colorCode + 'm+' + exports.humanize(this.diff) + '\u001b[0m');
  } else {
    args[0] = new Date().toISOString() + ' ' + name + ' ' + args[0];
  }
};

/**
 * Expose the debug namespace.
 */
exports.enable = common.enable;
exports.disable = common.disable;
exports.enabled = common.enabled;
exports.humanize = common.humanize;
exports.destroy = common.destroy;
exports.names = common.names;
exports.skips = common.skips;
exports.selectColor = common.selectColor;

/**
 * Active debug instances.
 */
exports.instances = common.instances;
