// debug/src/node.js
/**
 * src/node.js - Node.js-specific implementation of the debug module
 * 
 * This file provides the Node.js-specific functionality for the debug module,
 * including environment variable configuration, TTY detection, and output formatting.
 * It's essential for proper debug functionality in the Express.js server environment.
 * 
 * Copyright (c) 2014-2017 TJ Holowaychuk <tj@vision-media.ca>
 * Copyright (c) 2018-2021 Josh Junon <josh.junon@protonmail.com>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the 'Software'), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Module dependencies.
 */
const tty = require('tty');
const util = require('util');
const fs = require('fs');
const path = require('path');

/**
 * Cached process.stderr object to avoid multiple calls to `process.stderr`
 * @type {Object}
 */
let stderr = process.stderr;

/**
 * Map of ANSI color codes to their corresponding terminal color codes
 */
const colors = [
  6, 2, 3, 4, 5, 1 // cyan, green, yellow, blue, magenta, red
];

/**
 * Currently active debug mode, undefined if none
 * @type {String|undefined}
 */
let activeDebugMode;

/**
 * Previous timestamp for calculating time difference
 * @type {Number}
 */
let prevTime;

/**
 * Select a color for a debug namespace
 * @param {String} namespace The namespace string
 * @return {Number} An ANSI color code for the given namespace
 */
function selectColor(namespace) {
  let hash = 0;

  for (let i = 0; i < namespace.length; i++) {
    hash = ((hash << 5) - hash) + namespace.charCodeAt(i);
    hash |= 0; // Convert to 32bit integer
  }

  return colors[Math.abs(hash) % colors.length];
}

/**
 * Format the output message with timestamp, namespace, and data
 * @param {String} namespace The debug namespace
 * @param {String} message The message to format
 * @param {Object} options Formatting options
 * @return {String} Formatted output
 */
function formatArgs(namespace, message, options) {
  const useColors = options.useColors;
  const color = options.color;
  const diff = options.diff;
  
  let formatted = '';

  // Add timestamp
  const now = new Date();
  const timestamp = now.toISOString();
  formatted += timestamp + ' ';

  // Add namespace and color if enabled
  if (useColors) {
    formatted += '\u001B[3' + color + 'm' + namespace + '\u001B[0m ';
  } else {
    formatted += namespace + ' ';
  }

  // Add message
  formatted += message;

  // Add time diff
  if (diff) {
    formatted += ' +' + diff + 'ms';
  }

  return formatted;
}

/**
 * Load the DEBUG environment variable and process it
 * @return {Object} Object containing enabled namespaces
 */
function loadEnv() {
  const namespaces = process.env.DEBUG || '';
  const split = namespaces.split(/[\s,]+/);
  const enabledNamespaces = {};

  for (let i = 0; i < split.length; i++) {
    if (!split[i]) continue; // Skip empty strings
    const namespace = split[i].replace(/\*/g, '.*?');
    
    if (namespace[0] === '-') {
      // Negative namespace (exclude)
      enabledNamespaces[namespace.substr(1)] = false;
    } else {
      // Positive namespace (include)
      enabledNamespaces[namespace] = true;
    }
  }

  return enabledNamespaces;
}

/**
 * Check if a namespace is enabled
 * @param {String} namespace The namespace to check
 * @param {Object} enabledNamespaces Object containing enabled namespaces
 * @return {Boolean} Whether the namespace is enabled
 */
function isEnabled(namespace, enabledNamespaces) {
  if (process.env.DEBUG === '*') {
    return true;
  }

  // Check for direct match
  if (enabledNamespaces[namespace]) {
    return true;
  }

  // Check for wildcard matches
  for (const pattern in enabledNamespaces) {
    if (enabledNamespaces[pattern] && new RegExp('^' + pattern + '$').test(namespace)) {
      return true;
    }
  }

  return false;
}

/**
 * Create a debugger with the given namespace
 * @param {String} namespace The namespace for this debug instance
 * @return {Function} The debug function
 */
function createDebug(namespace) {
  // Check if TTY is available for colored output
  const useColors = Boolean(tty.isatty(stderr.fd) && process.env.TERM && process.env.TERM !== 'dumb');
  
  // Select color for this namespace
  const color = selectColor(namespace);

  // Return the debug function
  function debug(...args) {
    // If debugging is not enabled for this namespace, return immediately
    if (!debug.enabled) {
      return;
    }

    const self = debug;

    // Set the timestamp
    const curr = Date.now();
    const ms = curr - (prevTime || curr);
    prevTime = curr;

    // Format the message
    let message = util.format.apply(util, args);
    
    message = formatArgs(namespace, message, {
      useColors,
      color,
      diff: ms
    });

    // Output to stderr
    stderr.write(message + '\n');
  }

  debug.namespace = namespace;
  debug.enabled = false;
  debug.useColors = useColors;
  debug.color = color;
  debug.destroy = destroy;

  // Set enabled based on environment
  const enabledNamespaces = loadEnv();
  debug.enabled = isEnabled(namespace, enabledNamespaces);

  return debug;
}

/**
 * Destroys the debug instance
 * @return {Boolean} Whether the instance was destroyed
 */
function destroy() {
  activeDebugMode = undefined;
  return true;
}

/**
 * Enable a debug mode by namespaces
 * @param {String} namespaces The namespaces to enable (comma or space separated)
 * @return {void}
 */
function enable(namespaces) {
  activeDebugMode = namespaces;
  process.env.DEBUG = namespaces;
}

/**
 * Disable all debug modes
 * @return {void}
 */
function disable() {
  activeDebugMode = undefined;
  process.env.DEBUG = '';
}

/**
 * Returns all currently enabled namespaces
 * @return {String} Enabled namespaces
 */
function enabled() {
  return activeDebugMode || '';
}

/**
 * Load the DEBUG_COLORS environment variable
 * @return {Boolean} Whether colors should be used
 */
function useColors() {
  // Check if colors are explicitly disabled
  if (process.env.DEBUG_COLORS === '0' || process.env.DEBUG_COLORS === 'false') {
    return false;
  }
  
  // Check if we're in a TTY environment
  return Boolean(tty.isatty(stderr.fd) && process.env.TERM && process.env.TERM !== 'dumb');
}

/**
 * Save the debug output to a file
 * @param {String} path The file path to save to
 * @return {Function} Function to stop logging to file
 */
function toFile(filePath) {
  const stream = fs.createWriteStream(filePath, { flags: 'a' });
  const oldWrite = stderr.write;
  
  stderr.write = function(data) {
    oldWrite.apply(stderr, arguments);
    stream.write(data);
  };
  
  return function() {
    stderr.write = oldWrite;
    stream.end();
  };
}

/**
 * Initialize the module
 */
createDebug.enable = enable;
createDebug.disable = disable;
createDebug.enabled = enabled;
createDebug.useColors = useColors;
createDebug.toFile = toFile;
createDebug.destroy = destroy;

// Environment setup
if (process.env.DEBUG) {
  enable(process.env.DEBUG);
}

module.exports = createDebug;
