// depd/lib/browser.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module exports.
 * @public
 */

module.exports = depd

/**
 * Display a deprecation message in browser console.
 * This implementation is specifically designed for browser environments.
 *
 * @param {string} message
 * @param {object} [options]
 * @return {void}
 * @public
 */

function depd(message, options) {
  var opts = options || {}
  var header = opts.header || defaultHeader

  // Format the deprecation message
  var formattedMessage = header + ' ' + message

  // Get call site info if possible
  var stack = getStack()
  var callSite = getCallSite(stack)
  var callSiteString = callSite ? formatCallSite(callSite) : ''

  // Display the warning in the console
  if (typeof console !== 'undefined') {
    // Use console.warn if available (most browsers)
    if (console.warn) {
      console.warn(formattedMessage)
      if (callSiteString) {
        console.warn('    at ' + callSiteString)
      }
    } else if (console.log) {
      // Fallback to console.log if console.warn is not available
      console.log(formattedMessage)
      if (callSiteString) {
        console.log('    at ' + callSiteString)
      }
    }

    // Add trace if supported and not explicitly disabled
    if (console.trace && opts.trace !== false) {
      console.trace()
    }
  }

  return
}

/**
 * Default deprecation header.
 */

var defaultHeader = 'DeprecationWarning:'

/**
 * Get call site location from stack.
 * @private
 */

function getStack() {
  // Create an error to capture the stack
  var obj = {}
  var stack

  try {
    // Throw an error to generate a stack trace
    var err = new Error()
    Error.captureStackTrace
      ? Error.captureStackTrace(err, depd) // V8 (Chrome, Node.js)
      : (err = new Error()) // Other browsers

    stack = err.stack || ''
  } catch (e) {
    // Fallback for environments that might not support Error.stack
    stack = ''
  }

  return stack
}

/**
 * Extract call site from stack string.
 * @param {string} stack
 * @private
 */

function getCallSite(stack) {
  if (!stack) return null

  // Split the stack by lines
  var lines = stack.split('\n')

  // Find the first line that isn't part of depd
  for (var i = 0; i < lines.length; i++) {
    var line = lines[i]
    // Skip this module's functions
    if (line.indexOf('node_modules/depd') !== -1) {
      continue
    }
    // Skip empty lines and lines that don't look like call sites
    if (!line || line.indexOf('at ') !== 0) {
      continue
    }
    return line
  }

  return null
}

/**
 * Format a call site line from a stack trace.
 * @param {string} callSite
 * @private
 */

function formatCallSite(callSite) {
  // Remove the leading 'at ' prefix
  return callSite.replace(/^\s*at\s+/, '')
}

/**
 * Check if namespace is ignored.
 * @param {string} namespace
 * @private
 */

function isignored(namespace) {
  // Check for browser-specific ways to disable deprecation
  if (typeof window !== 'undefined' && window.noDeprecation) {
    return true
  }

  // Check localStorage if available
  try {
    if (typeof localStorage !== 'undefined' && localStorage.getItem('NO_DEPRECATION')) {
      var str = localStorage.getItem('NO_DEPRECATION') || ''
      return containsNamespace(str, namespace)
    }
  } catch (e) {
    // Ignore errors accessing localStorage
  }

  return false
}

/**
 * Check if namespace is traced.
 * @param {string} namespace
 * @private
 */

function istraced(namespace) {
  // Check for browser-specific ways to enable trace
  if (typeof window !== 'undefined' && window.traceDeprecation) {
    return true
  }

  // Check localStorage if available
  try {
    if (typeof localStorage !== 'undefined' && localStorage.getItem('TRACE_DEPRECATION')) {
      var str = localStorage.getItem('TRACE_DEPRECATION') || ''
      return containsNamespace(str, namespace)
    }
  } catch (e) {
    // Ignore errors accessing localStorage
  }

  return false
}

/**
 * Determine if namespace is contained in a string.
 * @param {string} str
 * @param {string} namespace
 * @private
 */

function containsNamespace(str, namespace) {
  var vals = str.split(/[ ,]+/)

  for (var i = 0; i < vals.length; i++) {
    if (vals[i] === '*' || vals[i] === namespace) {
      return true
    }
  }

  return false
}
