// depd/lib/caller.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module exports.
 * @public
 */

exports.callSiteToString = callSiteToString

/**
 * Convert a CallSite object to a string representation.
 * This function is used to generate a formatted string for deprecation warnings.
 *
 * @param {Object} callSite
 * @return {string}
 * @public
 */

function callSiteToString(callSite) {
  var fileName
  var fileLocation = ''
  var line
  var column
  var methodName
  var typeName

  // Get the call site properties
  fileName = callSite.getFileName() || '<anonymous>'
  line = callSite.getLineNumber()
  column = callSite.getColumnNumber()
  methodName = callSite.getFunctionName()
  typeName = callSite.getTypeName()

  // Format the file location
  fileLocation = fileName
  if (line !== undefined) {
    fileLocation += ':' + line
  }
  if (column !== undefined) {
    fileLocation += ':' + column
  }

  // Format the call site string
  var site = ''
  
  // Add the type name if available and different from the method name
  if (typeName && methodName && typeName + '.' !== methodName.substr(0, typeName.length + 1)) {
    site += typeName + '.'
  }

  // Add the method name if available
  if (methodName) {
    site += methodName
  }

  // Add constructor indication
  if (callSite.isConstructor()) {
    site += ' [constructor]'
  }

  // Add native function indication
  if (callSite.isNative()) {
    site += ' [native]'
  }

  // Add the file location
  if (site) {
    site += ' (' + fileLocation + ')'
  } else {
    site = fileLocation
  }

  return site
}

/**
 * Get the call site location as a string.
 * This function formats the call site with file, line, and column information.
 *
 * @param {Object} callSite
 * @return {string}
 * @private
 */

function callSiteLocation(callSite) {
  var fileName = callSite.getFileName() || '<anonymous>'
  var line = callSite.getLineNumber()
  var column = callSite.getColumnNumber()
  var formatted = fileName
  
  if (line) {
    formatted += ':' + line
  }
  
  if (column) {
    formatted += ':' + column
  }
  
  return formatted
}

/**
 * Get the call site file name.
 * This function returns the file name or a default value if not available.
 *
 * @param {Object} callSite
 * @return {string}
 * @private
 */

function callSiteFileName(callSite) {
  var fileName = callSite.getFileName()
  return fileName || '<anonymous>'
}

/**
 * Get the call site function name.
 * This function returns the function name or a default value if not available.
 *
 * @param {Object} callSite
 * @return {string}
 * @private
 */

function callSiteFunctionName(callSite) {
  var name = callSite.getFunctionName()
  
  if (!name) {
    // Try to get the method name if function name is not available
    name = callSite.getMethodName()
    
    if (!name) {
      // Use a placeholder for anonymous functions
      return '<anonymous>'
    }
  }
  
  return name
}
