// depd/lib/caller.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module exports.
 * @public
 */

module.exports = {
  callSiteToString: callSiteToString,
  getStack: getStack,
  parseCaller: parseCaller
}

/**
 * Convert a call site to a string representation.
 *
 * This function takes a V8 CallSite object and converts it to a string
 * representation that includes the file path, line number, and column.
 *
 * @param {Object} callSite - V8 CallSite object
 * @returns {string} String representation of call site
 * @public
 */
function callSiteToString(callSite) {
  var fileName
  var fileLocation = ''
  var methodName

  // Get the call site information
  if (callSite.isNative()) {
    fileLocation = 'native'
  } else if (callSite.isEval()) {
    fileName = callSite.getScriptNameOrSourceURL()
    if (!fileName) {
      fileLocation = callSite.getEvalOrigin()
    }
  } else {
    fileName = callSite.getFileName()
  }

  // Format the file location
  if (fileName) {
    fileLocation = fileName

    // Add line number if available
    var lineNumber = callSite.getLineNumber()
    if (lineNumber != null) {
      fileLocation += ':' + lineNumber

      // Add column number if available
      var columnNumber = callSite.getColumnNumber()
      if (columnNumber != null) {
        fileLocation += ':' + columnNumber
      }
    }
  }

  // Get method name if available
  var methodName = callSite.getFunctionName()
  if (!methodName) {
    methodName = callSite.getMethodName()

    // If we have a receiver, add it to the method name
    if (methodName) {
      var typeName = callSite.getTypeName()
      var isConstructor = callSite.isConstructor()

      if (typeName && !isConstructor) {
        methodName = typeName + '.' + methodName
      }
    }
  }

  // Return formatted call site string
  return methodName
    ? methodName + ' (' + fileLocation + ')'
    : fileLocation
}

/**
 * Get the call stack as an array of CallSite objects.
 *
 * This function captures the current stack trace and returns it as an array
 * of V8 CallSite objects, which can be used to determine where a function was called from.
 *
 * @returns {Array} Array of CallSite objects
 * @public
 */
function getStack() {
  var limit = Error.stackTraceLimit
  var obj = {}
  var prep = Error.prepareStackTrace

  // Capture the stack trace
  Error.prepareStackTrace = prepareObjectStackTrace
  Error.stackTraceLimit = Math.max(10, limit)
  Error.captureStackTrace(obj)

  // Get the stack
  var stack = obj.stack

  // Restore original stack trace limit and prepare function
  Error.prepareStackTrace = prep
  Error.stackTraceLimit = limit

  return stack
}

/**
 * Parse the caller from the stack.
 *
 * This function analyzes the stack trace to identify the caller of a function,
 * skipping specified number of frames to find the appropriate caller.
 *
 * @param {number} [skipFrames=0] - Number of frames to skip in the stack
 * @returns {Object|undefined} CallSite object for the caller or undefined if not found
 * @public
 */
function parseCaller(skipFrames) {
  skipFrames = skipFrames || 0

  // Get the call stack
  var stack = getStack()

  // Skip specified number of frames plus additional frames for this function
  // and the Error.captureStackTrace function itself
  var targetFrameIndex = skipFrames + 2

  // Return the appropriate caller
  return stack[targetFrameIndex]
}

/**
 * Prepare a stack trace for an object.
 *
 * This is a custom implementation of Error.prepareStackTrace that returns
 * the raw CallSite objects instead of a formatted string.
 *
 * @param {Object} obj - The object to prepare stack trace for
 * @param {Array} callSites - Array of CallSite objects
 * @returns {Array} The array of CallSite objects
 * @private
 */
function prepareObjectStackTrace(obj, callSites) {
  return callSites
}
