// depd/lib/compat.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var EventEmitter = require('events').EventEmitter
var util = require('util')

/**
 * Module exports.
 * @public
 */

module.exports = eventListenerCount
module.exports.EventEmitter = EventEmitter

/**
 * Node.js version detection and feature checking.
 * @private
 */

var nodeVersion = process.version.slice(1).split('.').map(Number)
var nodeVersionMajor = nodeVersion[0]
var nodeVersionMinor = nodeVersion[1]

/**
 * Determine if proper node.js EventEmitter.listenerCount exists.
 * @private
 */

var hasListenerCount = EventEmitter.listenerCount
  && typeof EventEmitter.listenerCount === 'function'

/**
 * Determine if node.js process.emitWarning exists.
 * @private
 */

var hasEmitWarning = typeof process.emitWarning === 'function'

/**
 * Determine if node.js process.emitWarning accepts options.
 * @private
 */

var hasEmitWarningOptions = hasEmitWarning && nodeVersionMajor >= 8

/**
 * Get the number of listeners for a given event.
 * @private
 */

function eventListenerCount (emitter, type) {
  return emitter.listeners(type).length
}

/**
 * Create a warning using process.emitWarning if available, otherwise
 * using console.warn.
 * @public
 */

module.exports.createWarning = function createWarning (message, type, code, ctor) {
  // Create warning object
  var warning = new Error(message)
  warning.name = type || 'DeprecationWarning'
  warning.code = code

  // Capture call site before emitting
  Error.captureStackTrace(warning, ctor || createWarning)

  // Emit warning
  emitWarning(warning)
}

/**
 * Emit a warning with the appropriate method.
 * @private
 */

function emitWarning (warning) {
  if (hasEmitWarningOptions) {
    // Use process.emitWarning with options object
    process.emitWarning(warning.message, {
      code: warning.code,
      name: warning.name,
      stack: warning.stack
    })
  } else if (hasEmitWarning) {
    // Use process.emitWarning with limited arguments
    process.emitWarning(warning.message, warning.name)
  } else if (warning.name === 'DeprecationWarning' && console.trace) {
    // Use console.trace for DeprecationWarning
    console.trace(warning.name + ': ' + warning.message)
  } else {
    // Use console.warn for all warnings
    console.warn(warning.name + ': ' + warning.message)
  }
}

/**
 * Create a getter function that will warn once about accessing a deprecated property.
 * @public
 */

module.exports.createGetter = function createGetter (obj, prop, message) {
  var warned = false

  return function getter () {
    if (!warned) {
      warned = true
      module.exports.createWarning(message, 'DeprecationWarning', undefined, getter)
    }

    return obj[prop]
  }
}

/**
 * Create a function that will warn once when called.
 * @public
 */

module.exports.createFunction = function createFunction (fn, message) {
  var warned = false

  return function deprecatedFunction () {
    if (!warned) {
      warned = true
      module.exports.createWarning(message, 'DeprecationWarning', undefined, deprecatedFunction)
    }

    return fn.apply(this, arguments)
  }
}

/**
 * Create a warning and invoke a function.
 * @public
 */

module.exports.createWarningFunction = function createWarningFunction (message, code) {
  return function () {
    module.exports.createWarning(message, 'DeprecationWarning', code)
  }
}

/**
 * Create a getter function that will warn once about accessing a deprecated property
 * and provide a replacement property.
 * @public
 */

module.exports.createReplaceGetter = function createReplaceGetter (obj, deprecated, replacement, message) {
  var warned = false

  return function replaceGetter () {
    if (!warned) {
      warned = true
      module.exports.createWarning(message, 'DeprecationWarning', undefined, replaceGetter)
    }

    return obj[replacement]
  }
}

/**
 * Create a function that will warn once when called and suggest a replacement function.
 * @public
 */

module.exports.createReplaceFunction = function createReplaceFunction (fn, deprecated, replacement, message) {
  var warned = false

  return function deprecatedFunction () {
    if (!warned) {
      warned = true
      module.exports.createWarning(message || deprecated + '() is deprecated. Use ' + replacement + '() instead.', 
        'DeprecationWarning', undefined, deprecatedFunction)
    }

    return fn.apply(this, arguments)
  }
}
