// depd/lib/index.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var callSiteToString = require('./caller').callSiteToString
var eventListenerCount = require('./compat').eventListenerCount
var hasEmitWarning = require('./compat').hasEmitWarning
var relative = require('path').relative

/**
 * Module exports.
 * @public
 */

module.exports = depd

/**
 * Get the path to base files on.
 */

var basePath = process.cwd()

/**
 * Create a deprecation function.
 *
 * @param {string} namespace
 * @param {object} [options]
 * @param {boolean} [options.stream=process.stderr]
 * @param {string} [options.header]
 * @return {function}
 * @public
 */

function depd(namespace, options) {
  if (!namespace) {
    throw new TypeError('argument namespace is required')
  }

  if (typeof namespace !== 'string') {
    throw new TypeError('argument namespace must be a string')
  }

  var opts = options || {}

  // Get environment-specific implementation
  var impl = process.env.NODE_ENV === 'production' || process.env.NO_DEPRECATION
    ? noop
    : process.browser
      ? require('./browser')
      : require('./node')

  function deprecate(message) {
    message = message || 'deprecated'

    // Determine if the message contains a namespace
    if (containsNamespace(message)) {
      message = message.replace(/::/, ' ')
    } else {
      message = namespace + ' ' + message
    }

    return impl(message, opts)
  }

  deprecate.function = wrapfunction
  deprecate.property = wrapproperty

  return deprecate
}

/**
 * Determine if namespace is separated.
 */

function containsNamespace(str) {
  var i = str.indexOf('::')
  return i !== -1 && i < str.length - 2
}

/**
 * Return a wrapped function in a deprecation message.
 *
 * This is a private function that is not exported directly.
 *
 * @param {Function} fn
 * @param {string} message
 * @returns {Function}
 * @private
 */

function wrapfunction(fn, message) {
  if (typeof fn !== 'function') {
    throw new TypeError('argument fn must be a function')
  }

  var args = new Array(arguments.length - 2)

  for (var i = 0; i < args.length; i++) {
    args[i] = arguments[i + 2]
  }

  var deprecatedfn = function () {
    this.deprecate(message)
    return fn.apply(this, arguments)
  }

  // Copy properties from original function
  for (var key in fn) {
    if (Object.prototype.hasOwnProperty.call(fn, key)) {
      deprecatedfn[key] = fn[key]
    }
  }

  deprecatedfn.name = fn.name
  deprecatedfn.toString = function () {
    return fn.toString()
  }

  return deprecatedfn
}

/**
 * Wrap property in a deprecation message.
 *
 * This is a private function that is not exported directly.
 *
 * @param {Object} obj
 * @param {string} prop
 * @param {string} message
 * @private
 */

function wrapproperty(obj, prop, message) {
  if (!obj || (typeof obj !== 'object' && typeof obj !== 'function')) {
    throw new TypeError('argument obj must be object')
  }

  var descriptor = Object.getOwnPropertyDescriptor(obj, prop)

  if (!descriptor) {
    throw new TypeError('cannot find property ' + prop)
  }

  // Create a new descriptor with deprecation warning
  var deprecatedDescriptor = {
    configurable: descriptor.configurable,
    enumerable: descriptor.enumerable
  }

  // Create deprecation message on access
  if (descriptor.get) {
    deprecatedDescriptor.get = function () {
      this.deprecate(message)
      return descriptor.get.apply(this, arguments)
    }
  }

  if (descriptor.set) {
    deprecatedDescriptor.set = function () {
      this.deprecate(message)
      return descriptor.set.apply(this, arguments)
    }
  }

  // Define the property with the modified descriptor
  Object.defineProperty(obj, prop, deprecatedDescriptor)
}

/**
 * No-op function for production environments.
 * @private
 */

function noop() {}
