// depd/lib/index.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var callSiteToString = require('./caller').callSiteToString
var eventListenerCount = require('./compat').eventListenerCount
var relative = require('path').relative

/**
 * Module exports.
 * @public
 */

module.exports = depd

/**
 * Get the path to base files on.
 */

var basePath = process.cwd()

/**
 * Determine if namespace is separated.
 */

function isSeparatedNamespace(str, sep) {
  if (typeof str !== 'string') {
    return false
  }

  var index = str.indexOf(sep)
  return index !== -1 && index !== 0 && index !== str.length - 1
}

/**
 * Get event listeners count.
 */

function getListenersCount(emitter, type) {
  return eventListenerCount(emitter, type)
}

/**
 * Create arguments string to keep arity.
 */

function createArgumentsString(arity) {
  var str = ''

  for (var i = 0; i < arity; i++) {
    str += ', arg' + i
  }

  return str.substr(2)
}

/**
 * Create stack string from stack.
 */

function createStackString(stack) {
  var str = this.name + ': ' + this.namespace

  if (this.message) {
    str += ' deprecated ' + this.message
  }

  for (var i = 0; i < stack.length; i++) {
    str += '\n    at ' + callSiteToString(stack[i])
  }

  return str
}

/**
 * Create deprecate for namespace in caller.
 */

function depd(namespace) {
  if (!namespace) {
    throw new TypeError('namespace required')
  }

  var stack = getStack()
  var site = callSiteLocation(stack[1])
  var file = site[0]

  function deprecate(message) {
    // Call to deprecated API
    var callSite = getStack()[0]
    var depSite = callSiteLocation(callSite)
    var relativeSite = (depSite[0] === file)
      ? depSite[1]
      : relative(basePath, depSite[0]) + ':' + depSite[1]

    // Determine if already warned
    var key = namespace + '|' + relativeSite
    if (warned[key]) {
      return
    }

    // Mark as warned
    warned[key] = true

    // Generate warning
    var msg = message
    var err = new Error()
    err.name = 'DeprecationWarning'
    err.namespace = namespace
    err.message = message

    // Capture stack
    Error.captureStackTrace(err, deprecate)

    // Format and output warning
    var formattedMessage = formatMessage(namespace, msg, relativeSite, callSite)
    emitWarning(formattedMessage, err)
  }

  function log() {
    // Call to deprecated API
    var callSite = getStack()[0]
    var depSite = callSiteLocation(callSite)
    var relativeSite = (depSite[0] === file)
      ? depSite[1]
      : relative(basePath, depSite[0]) + ':' + depSite[1]

    // Determine if already warned
    var key = namespace + '|' + relativeSite
    if (warned[key]) {
      return
    }

    // Mark as warned
    warned[key] = true

    // Generate warning
    var err = new Error()
    err.name = 'DeprecationWarning'
    err.namespace = namespace

    // Capture stack
    Error.captureStackTrace(err, log)

    // Format and output warning
    var formattedMessage = formatMessage(namespace, undefined, relativeSite, callSite)
    emitWarning(formattedMessage, err)
  }

  deprecate.function = wrapfunction
  deprecate.property = wrapproperty

  return deprecate
}

/**
 * Wrap property in a deprecation message.
 */

function wrapproperty(obj, prop, message) {
  if (!obj || (typeof obj !== 'object' && typeof obj !== 'function')) {
    throw new TypeError('obj must be object')
  }

  var descriptor = Object.getOwnPropertyDescriptor(obj, prop)

  if (!descriptor) {
    throw new TypeError('must call property on owner object')
  }

  if (!descriptor.configurable) {
    throw new TypeError('property must be configurable')
  }

  var deprecate = this
  var stack = getStack()
  var site = callSiteLocation(stack[1])

  // Set site name
  site.name = prop

  // Convert data descriptor
  if ('value' in descriptor) {
    descriptor = convertDataDescriptorToAccessor(obj, prop, message)
  }

  var get = descriptor.get
  var set = descriptor.set

  // Define getter
  if (typeof get === 'function') {
    descriptor.get = function getter() {
      return get.apply(this, arguments)
    }
  }

  // Define setter
  if (typeof set === 'function') {
    descriptor.set = function setter() {
      deprecate(message)
      return set.apply(this, arguments)
    }
  }

  Object.defineProperty(obj, prop, descriptor)
}

/**
 * Wrap function in a deprecation message.
 */

function wrapfunction(fn, message) {
  if (typeof fn !== 'function') {
    throw new TypeError('fn must be a function')
  }

  var args = createArgumentsString(fn.length)
  var stack = getStack()
  var site = callSiteLocation(stack[1])

  // Set site name
  site.name = fn.name

  var deprecatedfn = new Function('fn', 'log', 'deprecate', 'message', 'site',
    '"use strict"\n' +
    'return function (' + args + ') {\n' +
    '  log.call(deprecate, message, site)\n' +
    '  return fn.apply(this, arguments)\n' +
    '}'
  )(fn, log, this, message, site)

  return deprecatedfn
}

/**
 * Call site location getter.
 */

function callSiteLocation(callSite) {
  var file = callSite.getFileName() || '<anonymous>'
  var line = callSite.getLineNumber()
  var colm = callSite.getColumnNumber()

  if (callSite.isEval()) {
    file = callSite.getEvalOrigin() + ', ' + file
  }

  var site = [file, line, colm]

  site.callSite = callSite
  site.name = callSite.getFunctionName()

  return site
}

/**
 * Convert data descriptor to accessor descriptor.
 */

function convertDataDescriptorToAccessor(obj, prop, message) {
  var descriptor = Object.getOwnPropertyDescriptor(obj, prop)
  var value = descriptor.value

  descriptor.get = function getter() {
    return value
  }

  descriptor.set = function setter(val) {
    value = val
  }

  delete descriptor.value
  delete descriptor.writable

  return descriptor
}

/**
 * Create dynamic error with stack trace.
 */

function createError(name, msg, site) {
  var error = new Error(msg)
  var stack

  Object.defineProperty(error, 'name', {
    configurable: true,
    enumerable: false,
    value: name,
    writable: true
  })

  Object.defineProperty(error, 'stack', {
    configurable: true,
    enumerable: false,
    get: function () {
      if (stack !== undefined) {
        return stack
      }

      // Access stack to generate
      stack = createStackString.call(this, getStack())

      return stack
    },
    set: function setter(val) {
      stack = val
    }
  })

  return error
}

/**
 * Emit deprecation warning.
 */

function emitWarning(message, err) {
  if (process.emitWarning) {
    process.emitWarning(message, 'DeprecationWarning')
    return
  }

  if (process.stderr.isTTY) {
    // Format message
    var format = '\x1b[31;1mDeprecationWarning\x1b[0m: ' + message

    // Print to stderr
    process.stderr.write(format + '\n')
  } else {
    // Print to stderr
    process.stderr.write('DeprecationWarning: ' + message + '\n')
  }
}

/**
 * Format warning message.
 */

function formatMessage(namespace, message, site, callSite) {
  var callSiteName = callSite.getFunctionName()
  var depSiteName = site.name

  // Default message without call site info
  var msg = namespace + ' deprecated' + (message ? ' ' + message : '')

  // Add call site info
  if (callSite && depSiteName && callSiteName !== depSiteName) {
    msg += ' at ' + callSiteName + ' [' + site.join(':') + ']'
  }

  return msg
}

/**
 * Get call stack.
 */

function getStack() {
  var limit = Error.stackTraceLimit
  var obj = {}
  var prep = Error.prepareStackTrace

  Error.prepareStackTrace = prepareObjectStackTrace
  Error.stackTraceLimit = Math.max(10, limit)

  // Capture stack
  Error.captureStackTrace(obj)

  // Slice stack frames
  var stack = obj.stack.slice(1)

  Error.prepareStackTrace = prep
  Error.stackTraceLimit = limit

  return stack
}

/**
 * Prepare object stack trace.
 */

function prepareObjectStackTrace(obj, stack) {
  return stack
}

/**
 * Log function.
 */

function log(message, site) {
  message = message || ''

  var deprecate = this
  var stack = getStack()
  var caller = stack[0]

  if (site) {
    message = message + ' at ' + formatLocation(site)
  } else if (caller) {
    var callSite = callSiteLocation(caller)
    message = message + ' at ' + formatLocation(callSite)
  }

  // Generate warning
  var err = new Error()
  err.name = 'DeprecationWarning'
  err.namespace = deprecate.namespace
  err.message = message

  // Capture stack
  Error.captureStackTrace(err, log)

  // Format and output warning
  var formattedMessage = formatMessage(deprecate.namespace, message)
  emitWarning(formattedMessage, err)
}

/**
 * Format location.
 */

function formatLocation(callSite) {
  return relative(basePath, callSite[0]) +
    ':' + callSite[1] +
    ':' + callSite[2]
}

/**
 * Get the stack using Error.captureStackTrace.
 */

function getStack() {
  var limit = Error.stackTraceLimit
  var obj = {}
  var prep = Error.prepareStackTrace

  Error.prepareStackTrace = prepareObjectStackTrace
  Error.stackTraceLimit = Math.max(10, limit)

  // Capture stack
  Error.captureStackTrace(obj)

  // Slice stack frames
  var stack = obj.stack.slice(1)

  Error.prepareStackTrace = prep
  Error.stackTraceLimit = limit

  return stack
}

/**
 * Track for warned messages.
 */

var warned = Object.create(null)
