// depd/lib/node.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var EventEmitter = require('events').EventEmitter
var relative = require('path').relative
var callSiteToString = require('./caller').callSiteToString
var eventListenerCount = require('./compat').eventListenerCount

/**
 * Module exports.
 * @public
 */

module.exports = depd

/**
 * Get call site location as string.
 * @private
 */

function callSiteLocation(callSite) {
  var file = callSite.getFileName() || '<anonymous>'
  var line = callSite.getLineNumber()
  var colm = callSite.getColumnNumber()

  // Make relative paths more readable
  if (file.indexOf(process.cwd()) === 0) {
    file = relative(process.cwd(), file)
  }

  var site = callSite.getFunctionName()
    ? callSite.getFunctionName() + ' ('
    : ''

  site += file + ':' + line + ':' + colm

  site += callSite.getFunctionName()
    ? ')'
    : ''

  return site
}

/**
 * Get call site location from stack.
 * @private
 */

function getStack() {
  var limit = Error.stackTraceLimit
  var obj = {}
  var prep = Error.prepareStackTrace

  Error.prepareStackTrace = prepareObjectStackTrace
  Error.stackTraceLimit = Math.max(10, limit)

  // Capture the stack
  Error.captureStackTrace(obj)

  // Slice this function off the top
  var stack = obj.stack.slice(1)

  Error.prepareStackTrace = prep
  Error.stackTraceLimit = limit

  return stack
}

/**
 * Capture call site stack from v8.
 * @private
 */

function prepareObjectStackTrace(obj, stack) {
  return stack
}

/**
 * Display a deprecation message.
 * @param {string} message
 * @param {object} [options]
 * @return {void}
 * @public
 */

function depd(message, options) {
  var opts = options || {}
  var header = opts.header || defaultHeader

  // Get call site where deprecation was invoked
  var stack = getStack()
  var callSite
  for (var i = 0; i < stack.length; i++) {
    callSite = stack[i]
    // Skip this module's functions
    if (callSiteLocation(callSite).indexOf('node_modules/depd') !== -1) {
      continue
    }
    break
  }

  // Format the deprecation message
  var formattedMessage = message

  // Emit the deprecation warning
  if (process.emitWarning) {
    // Use the built-in Node.js warning system (Node.js v6+)
    var warning = new Error(formattedMessage)
    warning.name = 'DeprecationWarning'
    warning.code = 'DEPRECATED'

    // Capture the stack trace excluding this module's functions
    Error.captureStackTrace(warning, depd)

    try {
      process.emitWarning(warning)
    } catch (e) {
      // Fallback to console.error if emitWarning fails
      // (can happen in certain test environments or when process.stderr is not writable)
      console.error(header + ' ' + formattedMessage)
    }
  } else if (console.trace && opts.trace !== false) {
    // Fallback for older Node.js versions: use console.trace
    console.error(header + ' ' + formattedMessage)
    console.trace()
  } else {
    // Fallback for very old Node.js versions: simple console.error with call site
    console.error(header + ' ' + formattedMessage)
    if (callSite) {
      console.error('    at ' + callSiteToString(callSite))
    }
  }

  return
}

/**
 * Default deprecation header.
 */

var defaultHeader = 'DeprecationWarning:'

/**
 * Check if namespace is ignored.
 * @private
 */

function isignored(namespace) {
  // Process environment
  if (process.noDeprecation) {
    // Flag set in environment
    return true
  }

  var str = process.env.NO_DEPRECATION || ''

  // Namespace match
  return containsNamespace(str, namespace)
}

/**
 * Check if namespace is traced.
 * @private
 */

function istraced(namespace) {
  // Process environment
  if (process.traceDeprecation) {
    // Flag set in environment
    return true
  }

  var str = process.env.TRACE_DEPRECATION || ''

  // Namespace match
  return containsNamespace(str, namespace)
}

/**
 * Determine if namespace is contained in a string.
 * @private
 */

function containsNamespace(str, namespace) {
  var vals = str.split(/[ ,]+/)

  for (var i = 0; i < vals.length; i++) {
    if (vals[i] === '*' || vals[i] === namespace) {
      return true
    }
  }

  return false
}
