// depd/lib/node.js
/*!
 * depd
 * Copyright(c) 2014-2018 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var EventEmitter = require('events').EventEmitter
var relative = require('path').relative

/**
 * Module exports.
 * @public
 */

module.exports = depd

/**
 * Get the path to base files on.
 */

var basePath = process.cwd()

/**
 * Determine if namespace is already warned.
 * @private
 */

function isNamespaceWarned (namespace) {
  var warned = false

  // check if namespace is a constructor
  if (typeof namespace === 'function') {
    warned = isNamespaceWarned(namespace.prototype)

    if (warned) {
      return warned
    }
  }

  if (!namespace || typeof namespace !== 'object') {
    // early return for primitives to avoid throw from accessing Symbol
    return warned
  }

  var symbol = getDeprecationSymbol(namespace)

  if (symbol && namespace[symbol]) {
    warned = true
  }

  return warned
}

/**
 * Get deprecation symbol for namespace.
 * @private
 */

function getDeprecationSymbol (namespace) {
  if (!namespace) {
    return undefined
  }

  // use global symbol if exists
  if (typeof Symbol !== 'undefined') {
    return Symbol.for('namespace.deprecated')
  }

  // use pseudo-symbol
  return '@@namespace.deprecated'
}

/**
 * Create deprecation message.
 */

function createMessage (caller, message) {
  var stack = new Error().stack
  var callSite
  var i
  var site
  var match

  if (stack) {
    // get call site by removing leading lines and taking first line
    stack = stack.split('\n').slice(3)

    // find call site in stack
    for (i = 0; i < stack.length; i++) {
      match = stack[i].match(/\s*at\s*(.+?)\s*\((.+?):(\d+):(\d+)\)/)
      site = match && match[1] === caller && {
        name: match[1],
        file: match[2],
        line: parseInt(match[3], 10),
        column: parseInt(match[4], 10)
      }

      if (site) {
        callSite = site
        break
      }
    }
  }

  var file = callSite
    ? relative(basePath, callSite.file)
    : '<anonymous>'

  var line = callSite
    ? callSite.line
    : 1

  var column = callSite
    ? callSite.column
    : 1

  var location = file + ':' + line + ':' + column

  return message + ' (' + location + ')'
}

/**
 * Create deprecation.
 */

function depd (namespace) {
  if (!namespace) {
    throw new TypeError('namespace required')
  }

  function deprecate (message) {
    // warn for first call only
    if (isNamespaceWarned(namespace)) {
      return
    }

    // mark namespace as warned
    var symbol = getDeprecationSymbol(namespace)

    if (symbol) {
      namespace[symbol] = true
    }

    // generate message with call site info
    var caller = deprecate.caller
    var formatted = createMessage(caller && caller.name, message)

    // emit deprecation if listeners exist
    if (process.listeners('deprecation').length) {
      var err = new Error(formatted)
      err.name = 'DeprecationWarning'
      err.namespace = namespace
      err.message = message

      process.emit('deprecation', err)
      return
    }

    // use process.emitWarning if available
    if (typeof process.emitWarning === 'function') {
      process.emitWarning(formatted, 'DeprecationWarning')
      return
    }

    // fallback to standard error
    console.error('WARNING: ' + formatted)
  }

  deprecate._file = undefined
  deprecate._ignored = false
  deprecate._namespace = namespace
  deprecate._traced = false
  deprecate._warned = Object.create(null)

  deprecate.function = wrapfunction
  deprecate.property = wrapproperty

  return deprecate
}

/**
 * Wrap function in deprecation message.
 */

function wrapfunction (fn, message) {
  if (typeof fn !== 'function') {
    throw new TypeError('argument fn must be a function')
  }

  var args = new Array(arguments.length - 2)
  var caller = this

  // copy rest of arguments
  for (var i = 0; i < args.length; i++) {
    args[i] = arguments[i + 2]
  }

  return function wrapper () {
    caller(message)

    return fn.apply(this, arguments)
  }
}

/**
 * Wrap property in deprecation message.
 */

function wrapproperty (obj, prop, message) {
  if (!obj || (typeof obj !== 'object' && typeof obj !== 'function')) {
    throw new TypeError('argument obj must be object')
  }

  var descriptor = Object.getOwnPropertyDescriptor(obj, prop)

  if (!descriptor) {
    throw new TypeError('cannot find property ' + prop + ' on object')
  }

  // already warned
  if (descriptor.get && descriptor.get._deprecated) {
    return
  }

  // exit early if no accessor
  if (descriptor.configurable === false) {
    return
  }

  var caller = this
  var getter = descriptor.get
  var setter = descriptor.set

  // create new descriptor
  delete descriptor.get
  delete descriptor.set

  descriptor.get = function getter () {
    caller(message)
    return getter.apply(this, arguments)
  }

  descriptor.get._deprecated = true

  if (setter) {
    descriptor.set = function setter () {
      caller(message)
      return setter.apply(this, arguments)
    }
  }

  Object.defineProperty(obj, prop, descriptor)
}
