// statuses/index.js
/*!
 * statuses
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var codes = require('./codes.json')

/**
 * Module exports.
 * @public
 */

module.exports = status

// status code to message map
status.STATUS_CODES = codes

// array of status codes
status.codes = populateStatusesArray()

// status codes for redirects
status.redirect = {
  300: true,
  301: true,
  302: true,
  303: true,
  305: true,
  307: true,
  308: true
}

// status codes for empty bodies
status.empty = {
  204: true,
  205: true,
  304: true
}

// status codes for when you should retry the request
status.retry = {
  502: true,
  503: true,
  504: true
}

/**
 * Populate the statuses array for quick access.
 * @private
 */

function populateStatusesArray() {
  var arr = []

  Object.keys(codes).forEach(function forEachCode(code) {
    var message = codes[code]
    var status = Number(code)

    // Populate properties
    arr.push(status)
  })

  return arr
}

/**
 * Get the status code.
 *
 * Given a number, this will throw if it is not a known status
 * code, otherwise the code will be returned. Given a string,
 * the string will be parsed for a number and return the code
 * if valid, otherwise will lookup the code assuming this is
 * the status message.
 *
 * @param {string|number} code
 * @returns {number}
 * @public
 */

function status(code) {
  if (typeof code === 'number') {
    if (!codes[code]) throw new Error('invalid status code: ' + code)
    return code
  }

  if (typeof code !== 'string') {
    throw new TypeError('code must be a number or string')
  }

  // '403'
  var n = parseInt(code, 10)
  if (!isNaN(n)) {
    if (!codes[n]) throw new Error('invalid status code: ' + n)
    return n
  }

  // 'Forbidden'
  var status = getStatusCode(code)
  if (status) {
    return status
  }

  throw new Error('invalid status message: "' + code + '"')
}

/**
 * Get status code from message.
 * @private
 */

function getStatusCode(message) {
  var msg = message.toLowerCase()

  for (var code in codes) {
    if (codes[code].toLowerCase() === msg) {
      return Number(code)
    }
  }

  return null
}

/**
 * Check if a status code is a client error.
 * @public
 */

status.isClientError = function isClientError(code) {
  var num = status(code)
  return num >= 400 && num < 500
}

/**
 * Check if a status code is informational.
 * @public
 */

status.isInformational = function isInformational(code) {
  var num = status(code)
  return num >= 100 && num < 200
}

/**
 * Check if a status code is a redirect.
 * @public
 */

status.isRedirect = function isRedirect(code) {
  var num = status(code)
  return Boolean(status.redirect[num])
}

/**
 * Check if a status code is a server error.
 * @public
 */

status.isServerError = function isServerError(code) {
  var num = status(code)
  return num >= 500 && num < 600
}

/**
 * Check if a status code is a success.
 * @public
 */

status.isSuccess = function isSuccess(code) {
  var num = status(code)
  return num >= 200 && num < 300
}

/**
 * Check if a status code represents an empty response.
 * @public
 */

status.isEmpty = function isEmpty(code) {
  var num = status(code)
  return Boolean(status.empty[num])
}

/**
 * Check if a status code is retryable.
 * @public
 */

status.isRetryable = function isRetryable(code) {
  var num = status(code)
  return Boolean(status.retry[num])
}
