// statuses/index.js
/*!
 * statuses
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var codes = require('./codes.json')

/**
 * Module exports.
 * @public
 */

module.exports = status

// status code to message map
status.STATUS_CODES = codes

// array of status codes
status.codes = populateStatusesArray()

// status codes for redirects
status.redirect = {
  300: true,
  301: true,
  302: true,
  303: true,
  305: true,
  307: true,
  308: true
}

// status codes for empty bodies
status.empty = {
  204: true,
  205: true,
  304: true
}

// status codes for when you should retry the request
status.retry = {
  502: true,
  503: true,
  504: true
}

/**
 * Get the status code.
 *
 * Given a number, this will throw if it is not a known status code.
 * Given a string, this will throw if it cannot be parsed as a
 * numeric status code.
 *
 * @param {string|number} code
 * @returns {number}
 * @public
 */

function status (code) {
  if (typeof code === 'number') {
    if (!status[code]) throw new Error('invalid status code: ' + code)
    return code
  }

  if (typeof code !== 'string') {
    throw new TypeError('code must be a number or string')
  }

  // '403'
  var n = parseInt(code, 10)
  if (!isNaN(n)) {
    if (!status[n]) throw new Error('invalid status code: ' + n)
    return n
  }

  // 'Forbidden'
  var status_code = getStatusCodeByMessage(code)
  if (status_code) {
    return status_code
  }

  throw new Error('invalid status message: "' + code + '"')
}

/**
 * Populate the statuses array for status code lookup by number.
 * @private
 */

function populateStatusesArray () {
  var arr = []

  Object.keys(codes).forEach(function (code) {
    var n = parseInt(code, 10)
    arr.push(n)
    // store the code on the status function for status(code) lookup
    status[n] = n
  })

  return arr
}

/**
 * Get the status code from a message.
 * @private
 */

function getStatusCodeByMessage (message) {
  for (var code in codes) {
    if (codes[code].toLowerCase() === message.toLowerCase()) {
      return parseInt(code, 10)
    }
  }

  return null
}

/**
 * Check if a status code is a client error (4xx)
 * @public
 */

status.isClientError = function isClientError (statusCode) {
  var code = status(statusCode)
  return code >= 400 && code < 500
}

/**
 * Check if a status code is a server error (5xx)
 * @public
 */

status.isServerError = function isServerError (statusCode) {
  var code = status(statusCode)
  return code >= 500 && code < 600
}

/**
 * Check if a status code is an error (4xx or 5xx)
 * @public
 */

status.isError = function isError (statusCode) {
  var code = status(statusCode)
  return code >= 400 && code < 600
}

/**
 * Check if a status code is a redirect (3xx)
 * @public
 */

status.isRedirect = function isRedirect (statusCode) {
  var code = status(statusCode)
  return code >= 300 && code < 400
}

/**
 * Check if a status code is a success (2xx)
 * @public
 */

status.isSuccess = function isSuccess (statusCode) {
  var code = status(statusCode)
  return code >= 200 && code < 300
}

/**
 * Check if a status code is informational (1xx)
 * @public
 */

status.isInformational = function isInformational (statusCode) {
  var code = status(statusCode)
  return code >= 100 && code < 200
}

/**
 * Check if a status code indicates an empty body (204, 205, 304)
 * @public
 */

status.isEmpty = function isEmpty (statusCode) {
  var code = status(statusCode)
  return status.empty[code] === true
}

/**
 * Check if a status code is a retry (502, 503, 504)
 * @public
 */

status.isRetry = function isRetry (statusCode) {
  var code = status(statusCode)
  return status.retry[code] === true
}
