// express/lib/express.js
/**
 * Module dependencies.
 */

const EventEmitter = require('events').EventEmitter;
const mixin = require('merge-descriptors');
const proto = require('./application');
const Route = require('./router/route');
const Router = require('./router');
const req = require('./request');
const res = require('./response');

/**
 * Expose `createApplication()`.
 */

exports = module.exports = createApplication;

/**
 * Create an express application.
 *
 * @return {Function}
 * @api public
 */

function createApplication() {
  const app = function(req, res, next) {
    app.handle(req, res, next);
  };

  mixin(app, EventEmitter.prototype, false);
  mixin(app, proto, false);

  // expose the prototype that will get set on requests
  app.request = Object.create(req, {
    app: { configurable: true, enumerable: true, writable: true, value: app }
  });

  // expose the prototype that will get set on responses
  app.response = Object.create(res, {
    app: { configurable: true, enumerable: true, writable: true, value: app }
  });

  app.init();
  return app;
}

/**
 * Expose the prototypes.
 */

exports.application = proto;
exports.request = req;
exports.response = res;

/**
 * Expose constructors.
 */

exports.Route = Route;
exports.Router = Router;

/**
 * Expose middleware
 */

exports.json = require('body-parser').json;
exports.urlencoded = require('body-parser').urlencoded;

/**
 * Expose Router constructor as .Router
 */

exports.Router = Router;

/**
 * Create a new router.
 *
 * @return {Function}
 * @api public
 */

exports.router = function(options) {
  return new Router(options);
};

/**
 * Replace removed middleware with an appropriate error message.
 */

["bodyParser", "compress", "cookieSession", "session", "logger", "cookieParser",
  "favicon", "responseTime", "errorHandler", "timeout", "methodOverride", "vhost",
  "csrf", "directory", "limit", "multipart", "staticCache"].forEach(function (name) {
  Object.defineProperty(exports, name, {
    get: function () {
      throw new Error('Most middleware (like ' + name + ') is no longer bundled with Express and must be installed separately. Please see https://github.com/senchalabs/connect#middleware.');
    },
    configurable: true
  });
});

/**
 * Default engine extension map.
 *
 * @private
 */

const engines = exports.engines = {};

/**
 * Register template engine callback `fn`
 * for the given `ext`.
 *
 * @param {String} ext
 * @param {Function} fn
 * @return {app} for chaining
 * @api public
 */

exports.engine = function(ext, fn) {
  if (typeof ext !== 'string') {
    throw new Error('Extension must be a string');
  }
  if (ext[0] === '.') ext = ext.slice(1);
  engines[ext] = fn;
  return this;
};

/**
 * Expose static middleware
 */

exports.static = require('serve-static');
