// express/lib/utils.js
/**
 * Module dependencies.
 */

var Buffer = require('safe-buffer').Buffer;
var contentType = require('content-type');
var deprecate = require('depd')('express');
var flatten = require('array-flatten');
var mime = require('send').mime;
var etag = require('etag');
var proxyaddr = require('proxy-addr');
var qs = require('qs');
var querystring = require('querystring');

/**
 * Return strong ETag for `body`.
 *
 * @param {String|Buffer} body
 * @param {String} [encoding]
 * @return {String}
 * @api private
 */

exports.etag = createETagGenerator({ weak: false });

/**
 * Return weak ETag for `body`.
 *
 * @param {String|Buffer} body
 * @param {String} [encoding]
 * @return {String}
 * @api private
 */

exports.wetag = createETagGenerator({ weak: true });

/**
 * Check if `path` looks absolute.
 *
 * @param {String} path
 * @return {Boolean}
 * @api private
 */

exports.isAbsolute = function(path){
  if ('/' === path[0]) return true;
  if (':' === path[1] && ('\\' === path[2] || '/' === path[2])) return true; // Windows device path
  if ('\\\\' === path.substring(0, 2)) return true; // Windows UNC path
  return false;
};

/**
 * Flatten the given `arr`.
 *
 * @param {Array} arr
 * @return {Array}
 * @api private
 */

exports.flatten = deprecate.function(flatten,
  'utils.flatten: use array-flatten npm module instead');

/**
 * Normalize the given `type`, for example "html" becomes "text/html".
 *
 * @param {String} type
 * @return {Object}
 * @api private
 */

exports.normalizeType = function(type){
  return ~type.indexOf('/')
    ? acceptParams(type)
    : { value: mime.lookup(type), params: {} };
};

/**
 * Normalize `types`, for example "html" becomes "text/html".
 *
 * @param {Array} types
 * @return {Array}
 * @api private
 */

exports.normalizeTypes = function(types){
  var ret = [];

  for (var i = 0; i < types.length; ++i) {
    ret.push(exports.normalizeType(types[i]));
  }

  return ret;
};

/**
 * Generate Content-Type header appropriate for the type.
 * Non-ASCII type or parameter values will be encoded as per RFC 7231 sec. 3.1.1.2.
 *
 * @param {String|Object} type
 * @param {String|Object} [charset]
 * @return {String}
 * @api private
 */

exports.contentType = function contentTypeHelper(type, charset) {
  if (!type || typeof type !== 'string' && typeof type !== 'object') {
    throw new TypeError('argument type is required')
  }

  var parsed = typeof type === 'string'
    ? contentType.parse(type)
    : type

  if (charset) {
    parsed.parameters.charset = charset;
  }

  return contentType.format(parsed);
};

/**
 * Parse accept params `str` returning an
 * object with `.value` and `.params`.
 * For example "application/json;q=0.5" would
 * become `{ value: "application/json", params: { q: "0.5" } }`.
 *
 * @param {String} str
 * @return {Object}
 * @api private
 */

function acceptParams(str, tokens) {
  var parts = str.split(/ *; */);
  var ret = { value: parts[0], params: {} };

  for (var i = 1; i < parts.length; ++i) {
    var pms = parts[i].split(/ *= */);
    if (pms.length !== 2) {
      continue;
    }
    ret.params[pms[0]] = pms[1].replace(/^"([^"]*)"$/, '$1');
  }

  return ret;
}

/**
 * Compile "etag" value to function.
 *
 * @param  {Boolean|String|Function} val
 * @return {Function}
 * @api private
 */

exports.compileETag = function(val) {
  var fn;

  if (typeof val === 'function') {
    return val;
  }

  switch (val) {
    case true:
    case 'weak':
      fn = exports.wetag;
      break;
    case false:
      break;
    case 'strong':
      fn = exports.etag;
      break;
    default:
      throw new TypeError('unknown value for etag function: ' + val);
  }

  return fn;
};

/**
 * Compile "query parser" value to function.
 *
 * @param  {String|Function} val
 * @return {Function}
 * @api private
 */

exports.compileQueryParser = function compileQueryParser(val) {
  var fn;

  if (typeof val === 'function') {
    return val;
  }

  switch (val) {
    case true:
    case 'simple':
      fn = querystring.parse;
      break;
    case false:
      fn = newObject;
      break;
    case 'extended':
      fn = parseExtendedQueryString;
      break;
    default:
      throw new TypeError('unknown value for query parser function: ' + val);
  }

  return fn;
};

/**
 * Compile "proxy trust" value to function.
 *
 * @param  {Boolean|String|Number|Array|Function} val
 * @return {Function}
 * @api private
 */

exports.compileTrust = function(val) {
  if (typeof val === 'function') return val;

  if (val === true) {
    // Support plain true/false
    return function(){ return true };
  }

  if (typeof val === 'number') {
    // Support trusting hop count
    return function(a, i){ return i < val };
  }

  if (typeof val === 'string') {
    // Support comma-separated values
    val = val.split(',').map(function (v) { return v.trim() });
  }

  return proxyaddr.compile(val || []);
};

/**
 * Set the charset in a given Content-Type string.
 *
 * @param {String} type
 * @param {String} charset
 * @return {String}
 * @api private
 */

exports.setCharset = function setCharset(type, charset) {
  if (!type || !charset) {
    return type;
  }

  // parse type
  var parsed = contentType.parse(type);

  // set charset
  parsed.parameters.charset = charset;

  // format type
  return contentType.format(parsed);
};

/**
 * Create an ETag generator function, generating ETags with
 * the given options.
 *
 * @param {object} options
 * @return {function}
 * @private
 */

function createETagGenerator (options) {
  return function generateETag (body, encoding) {
    var buf = !Buffer.isBuffer(body)
      ? Buffer.from(body, encoding)
      : body

    return etag(buf, options)
  }
}

/**
 * Parse an extended query string with qs.
 *
 * @param {String} str
 * @return {Object}
 * @private
 */

function parseExtendedQueryString(str) {
  return qs.parse(str, {
    allowPrototypes: true
  });
}

/**
 * Return new empty object.
 *
 * @return {Object}
 * @api private
 */

function newObject() {
  return {};
}

/**
 * Normalize the path string,
 * returning a regular expression.
 *
 * Inspired by pathtoRegexp 'normalize' method.
 *
 * @param  {string} path
 * @param  {boolean} strict
 * @return {string}
 */

exports.normalizePath = function normalizePath(path, strict) {
  if (!path) return path;
  
  // Ensure path starts with a slash
  if (path[0] !== '/') {
    path = '/' + path;
  }
  
  // Remove trailing slash for non-root paths
  if (!strict && path.length > 1 && path[path.length - 1] === '/') {
    path = path.substr(0, path.length - 1);
  }
  
  return path;
};

/**
 * Decode a URL component safely, handling malformed URLs
 * and preventing security issues with decodeURIComponent.
 *
 * @param {string} component - URL component to decode
 * @return {string} Decoded component or original if invalid
 * @api private
 */
exports.safeDecodeURIComponent = function safeDecodeURIComponent(component) {
  try {
    return decodeURIComponent(component);
  } catch (err) {
    return component;
  }
};

/**
 * Check if a value is an object (excluding null).
 *
 * @param {*} val - Value to check
 * @return {boolean} True if object, false otherwise
 * @api private
 */
exports.isObject = function isObject(val) {
  return val !== null && typeof val === 'object' && !Array.isArray(val);
};

/**
 * Merge objects together, with later objects
 * overriding properties in earlier ones.
 *
 * @param {...Object} objs - Objects to merge
 * @return {Object} Merged object
 * @api private
 */
exports.merge = function merge() {
  var target = {};
  
  for (var i = 0; i < arguments.length; i++) {
    var source = arguments[i];
    
    if (source) {
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
  }
  
  return target;
};

/**
 * Escape a string for use in a regular expression.
 *
 * @param {string} str - String to escape
 * @return {string} Escaped string
 * @api private
 */
exports.escapeRegExp = function escapeRegExp(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
};

/**
 * Check if a value is a string.
 *
 * @param {*} val - Value to check
 * @return {boolean} True if string, false otherwise
 * @api private
 */
exports.isString = function isString(val) {
  return typeof val === 'string';
};

/**
 * Check if a HTTP status code is a redirect code.
 *
 * @param {number} code - Status code to check
 * @return {boolean} True if redirect, false otherwise
 * @api private
 */
exports.isRedirectStatus = function isRedirectStatus(code) {
  return [301, 302, 303, 305, 307, 308].indexOf(code) !== -1;
};

/**
 * Strip content-* header fields.
 *
 * @param {Object} headers
 * @return {Object} Modified headers object
 * @api private
 */
exports.stripContentHeaders = function stripContentHeaders(headers) {
  var newHeaders = {};
  
  for (var header in headers) {
    if (!header.toLowerCase().startsWith('content-')) {
      newHeaders[header] = headers[header];
    }
  }
  
  return newHeaders;
};
