// http-errors/index.js
/*!
 * http-errors
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var deprecate = require('depd')('http-errors')
var setPrototypeOf = require('setprototypeof')
var statuses = require('statuses')
var inherits = require('inherits')
var toIdentifier = require('toidentifier')

/**
 * Module exports.
 * @public
 */

module.exports = createError
module.exports.HttpError = createHttpErrorConstructor()
module.exports.isHttpError = createIsHttpErrorFunction(module.exports.HttpError)

// Populate exports for all constructors
populateConstructorExports(module.exports, statuses.codes, module.exports.HttpError)

/**
 * Get the code class of a status code.
 * @private
 */

function codeClass (status) {
  return Number(String(status).charAt(0) + '00')
}

/**
 * Create a new HTTP Error.
 *
 * @param {number} status Status code.
 * @param {string} message Error message.
 * @param {object} [props] Additional properties to include on the error.
 * @return {Error} The error object.
 * @public
 */

function createError () {
  // Normal call signature
  var err
  var msg
  var status = 500
  var props = {}
  var args = Array.prototype.slice.call(arguments)

  for (var i = 0; i < args.length; i++) {
    var arg = args[i]

    if (arg instanceof Error) {
      err = arg
      status = err.status || err.statusCode || status
      continue
    }

    switch (typeof arg) {
      case 'string':
        msg = arg
        break
      case 'number':
        status = arg
        if (i !== 0) {
          deprecate('non-first-argument status code; use createError(' + arg + ', ...)')
        }
        break
      case 'object':
        props = arg
        break
    }
  }

  if (typeof status === 'number' && (status < 400 || status >= 600)) {
    deprecate('non-error status code; use only 4xx or 5xx status codes')
  }

  if (typeof status !== 'number' ||
    (!statuses[status] && (status < 400 || status >= 600))) {
    status = 500
  }

  // Create error
  var HttpError = createError[status] || createError[codeClass(status)]
  var message = msg || statuses[status] || String(status)

  if (!err) {
    // Create error
    err = HttpError
      ? new HttpError(message)
      : new Error(message)

    Error.captureStackTrace(err, createError)
  }

  // Set status code
  err.status = err.statusCode = status

  // Set properties
  for (var key in props) {
    if (key !== 'status' && key !== 'statusCode') {
      err[key] = props[key]
    }
  }

  return err
}

/**
 * Create HTTP error constructor.
 * @private
 */

function createHttpErrorConstructor () {
  function HttpError () {
    // Create the error object
    var message = arguments.length > 0
      ? arguments[0]
      : statuses[this.status]
    var err = new Error(message)

    // Capture a stack trace
    Error.captureStackTrace(err, this.constructor)

    // Set the prototype
    setPrototypeOf(err, this.constructor.prototype)

    // Assign status properties
    this.status = this.statusCode = this.constructor.status
    this.expose = this.constructor.expose

    // Add properties to the error instance
    for (var i = 0; i < arguments.length; i++) {
      if (typeof arguments[i] === 'object') {
        var keys = Object.keys(arguments[i])
        for (var j = 0; j < keys.length; j++) {
          var key = keys[j]
          if (key !== 'status' && key !== 'statusCode') {
            err[key] = arguments[i][key]
          }
        }
      }
    }

    return err
  }

  inherits(HttpError, Error)

  return HttpError
}

/**
 * Create isHttpError function for HttpError
 * @private
 */

function createIsHttpErrorFunction (HttpError) {
  return function isHttpError (val) {
    if (!val || typeof val !== 'object') {
      return false
    }

    if (val instanceof HttpError) {
      return true
    }

    return val instanceof Error &&
      typeof val.status === 'number' &&
      typeof val.statusCode === 'number' &&
      val.status === val.statusCode
  }
}

/**
 * Create a constructor for a client error.
 * @private
 */

function createClientErrorConstructor (HttpError, name, code) {
  var className = toClassName(name)

  function ClientError (message) {
    // Create the error object
    var err = new HttpError(message)

    // Capture a stack trace
    Error.captureStackTrace(err, this.constructor)

    // Set the prototype
    setPrototypeOf(err, this.constructor.prototype)

    return err
  }

  inherits(ClientError, HttpError)

  ClientError.prototype.name = className
  ClientError.prototype.statusCode = ClientError.prototype.status = code
  ClientError.prototype.expose = true

  return ClientError
}

/**
 * Create a constructor for a server error.
 * @private
 */

function createServerErrorConstructor (HttpError, name, code) {
  var className = toClassName(name)

  function ServerError (message) {
    // Create the error object
    var err = new HttpError(message)

    // Capture a stack trace
    Error.captureStackTrace(err, this.constructor)

    // Set the prototype
    setPrototypeOf(err, this.constructor.prototype)

    return err
  }

  inherits(ServerError, HttpError)

  ServerError.prototype.name = className
  ServerError.prototype.statusCode = ServerError.prototype.status = code
  ServerError.prototype.expose = false

  return ServerError
}

/**
 * Populate the exports object with constructors for every error class.
 * @private
 */

function populateConstructorExports (exports, codes, HttpError) {
  codes.forEach(function forEachCode (code) {
    var CodeError
    var name = toIdentifier(statuses[code])

    switch (codeClass(code)) {
      case 400:
        CodeError = createClientErrorConstructor(HttpError, name, code)
        break
      case 500:
        CodeError = createServerErrorConstructor(HttpError, name, code)
        break
    }

    if (CodeError) {
      // Export the constructor
      exports[code] = CodeError
      exports[name] = CodeError
    }
  })

  // Add the base error class
  exports.HttpError = HttpError
}

/**
 * Convert a string of words to a JavaScript identifier.
 * @private
 */

function toClassName (name) {
  return name.split(' ').map(function (token) {
    return token.slice(0, 1).toUpperCase() + token.slice(1)
  }).join('').replace(/[^ _0-9a-z]/gi, '')
}
